---
title: "1100-07212020-expanded-samples"
author: "Mac Campbell"
date: "July 21, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
library(tidyverse)
```

## Finding the intersect of the big data 
```{r}
data<-read_csv("metadata/skt_lhp_samples-07212020.csv")  %>% 
  rename(Sample_ID=full_id) %>% rename(Phenotype=`supervised assignments`)

#A couple of things, - and _ as delimiter
#There are no leading zeros in one of our lists

data<-data %>% separate(Sample_ID, sep = "_", into=c("Sample Year","Survey","ID"))
data$ID<-as.numeric(data$ID)

data<- data %>% mutate(Sample_ID = paste(`Sample Year`, Survey, ID, sep="-"))

seqs<-read_csv("metadata/Field Ids.csv") %>% rename(Sample_ID=`Field ID`)

meta<-left_join(seqs,data) %>% filter(Phenotype %in% c("FWR","MIG"))

meta %>% select(Phenotype) %>% group_by(Phenotype) %>% summarize(Count=n())
```

That doesn't seem so great.
```{r}
data %>% select(`Sample Year`, Survey) %>% group_by(`Sample Year`, Survey) %>% summarize(Count=n())
```

```{sh}
grep SKT metadata/meta.csv  | wc -l
```

What about the distribution of seqs we have?

```{r}
 seqs %>% select(Sample_ID) %>% separate(Sample_ID, sep = "-", into=c("Sample Year","Survey","ID")) %>% group_by(`Sample Year`, Survey) %>% summarize(Count=n())
```

###What to do?

It looks like to me that the SKT2005-Plate3	05_168 fields may beuseful.

Is there a 2005-SKT-168?

```{r}
data %>% filter(Sample_ID=="2005-SKT-168")
```
I'll make a separate table
`Macs-MBP-2:metadata mac$ grep SKT meta.csv  > SKT.meta-no-field-ids.csv`

```{r}
skt<-read_csv("metadata/SKT.meta-no-field-ids.csv") %>% separate(`Sample ID`, sep="_|-",
                                                                 into=c("Year","Ind"))
skt<-skt %>% mutate(SamplingYear=`Birth year` + 1)
skt<-skt %>% mutate(Sample_ID = paste(SamplingYear,"SKT",Ind, sep="-")) %>%
  rename(`Sequence File Name` = filename)

meta2<-left_join(skt,data) %>% filter(Phenotype %in% c("FWR","MIG"))

```

###Now to merge

I should be able to downsample the data to a few interesting columns.

```{r}
#Drop the two bad samples I know about
#Ht20-30_2012_F04
#Ht20-77_2012_E10
metasub<-meta %>% filter(!`Sequence File Name` %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10")) %>%
            select(`Sequence File Name`, Sample_ID, Phenotype, `Birth Year (HOBBS)`) %>%
  rename(BirthYear=`Birth Year (HOBBS)`)
metasub$BirthYear<-as.numeric(metasub$BirthYear)
meta2sub<-meta2 %>% select(`Sequence File Name`, Sample_ID, Phenotype, `Birth year`) %>% rename(BirthYear=`Birth year`)

sub<-bind_rows(metasub, meta2sub)

bams<-select(sub, `Sequence File Name`) %>% mutate(Path=paste0("bams/",`Sequence File Name`,".sort-n.fixmate-m.sort.markdup-r.bam")) %>% select(Path)
write_tsv(bams, path = "bamlists/expanded.bamlist", col_names = FALSE)

phenos<-select(sub,Phenotype)
phenos$Phenotype<-gsub("FWR",0,phenos$Phenotype)
phenos$Phenotype<-gsub("MIG",1,phenos$Phenotype)
write_tsv(phenos, path = "phenos/expanded.phenos", col_names=FALSE)

#What did we end up with?
sub %>% group_by(Phenotype, BirthYear) %>% summarize(Count=n())
sub %>% group_by(Phenotype) %>% summarize(Count=n())

```

#Make a PCA 
And run association testing at 90% missing.
402 individuals!!

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/expanded.bamlist \
-minInd 362 -GL 1 -out $HOME/ds-lh/outputs/1100/expanded-pca \
-doGlf 2  -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-6 -minMapQ 30 -minQ 20 > outputs/1100/expanded-pca.out 2> outputs/1100/expanded-pca.err &

#Generate Covariance Matrix
python $HOME/pcangsd/pcangsd.py -beagle $HOME/ds-lh/outputs/1200/expanded-pca.beagle.gz -admix -o $HOME/ds-lh/outputs/1100/expanded-pca

##Association testing without imputation
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/expanded.bamlist \
-yBin $HOME/ds-lh/phenos/expanded.phenos -minMapQ 30 -minQ 20 -minInd 362 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1100/assoc -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/ds-lh/genome/Hypomesus-transpacificus_10X_F_A.pseudohap2.1.fasta > outputs/1100/assoc.out 2> outputs/1100/assoc.err &

```
