---
title: "1200-new-genome"
author: "Mac Campbell"
date: "2/16/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

# New Genome
/home/maccamp/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa    

## Re-align Data
/group/millermrgrp2/shannon/projects/DS_history/data     
Moving to data/ and using 2012.bamlist to find the right sequences.

```{r}
bams<-read_tsv("bamlists/2012.bamlist", col_names="Bam")
bams$Sample<-gsub("bams/|.sort-n.fixmate-m.sort.markdup-r.bam","",bams$Bam)
#Need form Ht19-89_2012_A12_R1.fastq Ht19-89_2012_A12_R2.fastq
bams<-bams %>% mutate(R1=paste0(Sample,"_R1")) %>% mutate(R2=paste0(Sample,"_R2")) %>% 
  mutate(Final=paste0("bams/",R1,".sort.flt.bam"))

align<-bams %>% select(R1, R2)
write_tsv(align, "bamlists/2012-samples.tsv", col_names = FALSE)

new2012<-bams %>% select(Final)
write_tsv(new2012, "bamlists/2012-newgenome.bamlist", col_names = FALSE)

```

Now, we should be able to run an alignment to the new reference.    

`maccamp@farm:~/ds-lh/data$ ./doAlign.sh 2012-samples.tsv $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa`    
Let's do  these things for our 2012 data 

__1__ PCA     

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Some shameless copying from Eric A.
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}
```


```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome.bamlist \
-minInd 100 -GL 1 -out $HOME/ds-lh/outputs/1200/pca \
-doGlf 2  -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-6 -minMapQ 20 -minQ 20 > outputs/1200/pca.out 2> outputs/1200/pca.err &

#Generate Covariance Matrix
python $HOME/pcangsd/pcangsd.py -beagle $HOME/ds-lh/outputs/1200/pca.beagle.gz -admix -o $HOME/ds-lh/outputs/1200/pca
```


```{r}
pca_meta<-read_csv("metadata/metadata-2012.csv") %>% rename(sample=`Sequence File Name`) 
cov<-read_delim("outputs/1200/pca.cov", delim=" ", col_names=FALSE) %>% as.matrix()

pca <- covar2pcs(pca_meta$sample, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pca_meta, by = "sample")

#drop outliers, deleted from bamlist and rerun
#Ht20-30_2012_F04
#Ht20-77_2012_E10
pp_meta<-pp_meta %>% filter(!(sample %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10")))
ggplot(pp_meta, aes(x = val_x, y = val_y, fill=Phenotype), alpha=0.75) +
  geom_point(pch = 21, size = 2) +
  scale_fill_discrete(na.value = "white") + 
  scale_fill_manual(values = c("red","blue")) +
  facet_grid(PCy ~ PCx, scales = "free")+
  theme_bw()

ggsave("outputs/1200/6-6-pca.pdf", width=14, height=12)

```


__2__ Association testing    
Phenotype file is in order.
```{sh,eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome.bamlist \
-yBin $HOME/ds-lh/phenos/2012.phenos -minMapQ 20 -minQ 20 -minInd 100 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1200/assoc -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1200/assoc.out 2> outputs/1200/assoc.err &

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome.bamlist \
-yBin $HOME/ds-lh/phenos/2012.phenos -minMapQ 20 -minQ 20 -minInd 67 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1200/assoc-half -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1200/assoc-half.out 2> outputs/1200/assoc-half.err &

```

75% thresh:     
lg02	11230311	A	C	0.309547	13.329619      
lg17	4018288	T	A	0.037558	13.564628     
lg26	9000090	A	G	0.061632	13.666326    
lg15	2268817	C	T	0.236092	13.810386    
lg14	14334932	G	T	0.049173	14.552699     
lg02	11230464	T	G	0.303847	15.540977    
lg19	473790	C	G	0.095084	15.708995     
lg22	14194108	G	A	0.203564	16.577111    
lg01	2467271	G	A	0.091513	16.836151    
lg17	14078225	C	T	0.063486	17.638608    

50% thresh:    
lg02	11230311	A	C	0.309547	13.329619   
lg17	4018288	T	A	0.037558	13.564628   
lg26	9000090	A	G	0.061632	13.666326   
lg15	2268817	C	T	0.236092	13.810386   
lg14	14334932	G	T	0.049173	14.552699   
lg02	11230464	T	G	0.303847	15.540977   
lg19	473790	C	G	0.095084	15.708995   
lg22	14194108	G	A	0.203564	16.577111   
lg01	2467271	G	A	0.091513	16.836151    
lg17	14078225	C	T	0.063486	17.638608     


Imputation version.    
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -GL 1 -minMapQ 20 -minQ 20 -minInd 67 -out $HOME/ds-lh/outputs/1200/input -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1  -bam $HOME/ds-lh/bamlists/2012-newgenome.bamlist -doGlf 2 >outputs/1200/impute1.out 2>outputs/1200/impute1.err &


#Using version 3.3.2 for compatibility with ANGSD
srun -p high -t 12:00:00 --mem=32G --nodes=2 java -Xmx15000m -jar $HOME/beagle/beagle.jar like=$HOME/ds-lh/outputs/1200/input.beagle.gz out=$HOME/ds-lh/outputs/1200/beagleOut

#Notes Warning: All data sets have > 0.07 missing alleles for the markers in the markers file.
#         To obtain the most accurate imputation, at least one input Beagle file should be genotyped
#         for all markers in the markers file and should have < 0.07 missing alleles.
# 98353 markers

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -doMaf 4 -beagle $HOME/ds-lh/outputs/1200/beagleOut.input.beagle.gz.gprobs.gz -fai $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa.fai  -yBin $HOME/ds-lh/phenos/2012.phenos -doAsso 2 -out $HOME/ds-lh/outputs/1200/impute-assoc
```
impute-assoc.lrt0.gz keeps coming up with unexpected end of file. It is actually empty.    

Reducing missing data...

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -GL 1 -minMapQ 20 -minQ 20 -minInd 127 -out $HOME/ds-lh/outputs/1200/input2 -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1  -bam $HOME/ds-lh/bamlists/2012-newgenome.bamlist -doGlf 2 >outputs/1200/impute2.out 2>outputs/1200/impute2.err &

srun -p high -t 12:00:00 --mem=32G --nodes=2 java -Xmx15000m -jar $HOME/beagle/beagle.jar like=$HOME/ds-lh/outputs/1200/input2.beagle.gz out=$HOME/ds-lh/outputs/1200/beagleOut2

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -doMaf 4 -beagle $HOME/ds-lh/outputs/1200/beagleOut2.input2.beagle.gz.gprobs.gz -fai $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa.fai  -yBin $HOME/ds-lh/phenos/2012.phenos -doAsso 2 -out $HOME/ds-lh/outputs/1200/impute-assoc2
```

Now, let's compute the number of reads mapped for our samples and keep those that are higher (reducing sample sizes, but, hey, maybe that's the right tack.)

### Get number of reads/bam file.    

` ls | grep sort.flt.bam | grep -v bai | while read line; do samtools flagstat $line | grep mapped | head -n 1 >> counts.txt; done;`
` ls | grep sort.flt.bam | grep -v bai >> counts.files.txt`

```{r}
files<-read_tsv("outputs/1200/counts.files.txt", col_names="File")
counts<-read_tsv("outputs/1200/counts.txt", col_names="Counts")
counts$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts$Counts, fixed = TRUE)
phenos<-read_tsv("phenos/2012.phenos", col_name="Pheno")
comb<-bind_cols(files, counts)

bams$File<-gsub("bams/", "",bams$Final)

bams<-bams %>% left_join(comb)
bams<-bind_cols(bams, phenos)
```

Now to filter.

```{r}
bams$Counts<-as.numeric(bams$Counts)
```

```{r}
ggplot(bams) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Pheno)
```

```{r}
tops<-bams %>% top_frac(.5, Counts)
tops %>% group_by(Pheno) %>% summarize(Count=n())
ggplot(tops) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Pheno) +
  xlim(0, max(bams$Counts))
```


Create new bamlist and pheno list.
```{r}
new05<-tops %>% select(Final)
write_tsv(new05, "bamlists/2012-newgenome05.bamlist", col_names = FALSE)

new05pheno<-tops %>% select(Pheno)
write_tsv(new05pheno, "phenos/2012-newgenome05.phenos", col_names = FALSE)
```

Association testing, 67 individuals 75% thresh, 50.25:
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome05.bamlist \
-yBin $HOME/ds-lh/phenos/2012-newgenome05.phenos -minMapQ 20 -minQ 20 -minInd 51 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1200/assoc05 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1200/assoc05.out 2> outputs/1200/assoc05.err &
```
lg01	2467271	G	A	0.077797	13.031384   
lg23	5832618	A	T	0.113305	13.124640   
lg09	7338206	C	T	0.149206	13.275099    
lg11	11148662	G	A	0.074642	13.720920     
scaffold_314	6295	C	A	0.074632	13.721667     
lg21	47867	C	T	0.279261	13.899696    
scaffold_314	6080	A	T	0.076713	14.010966     
lg10	13089180	A	C	0.078174	14.150752    
lg02	11230311	A	C	0.293849	21.220818     
lg02	11230464	T	G	0.317309	26.532230     

Let's check out lg02.

```{r}
df<-read_tsv(file="outputs/1200/assoc05.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf")

arrange(df, -log10p) %>% top_n(20)
df2 <- filter(df, grepl("lg",Chromosome))
highest<-df2 %>% top_frac(.001,log10p) 

ggplot(df2) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=highest, aes(x=Position,y=log10p), pch=21, fill="cyan") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=10, scales="free_x") +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

```

What about significance thresholds?
https://bmcgenomics.biomedcentral.com/track/pdf/10.1186/s12864-019-6324-7.pdf   
"Bonferroni test P < 0.05/n or P < 0.01/n (n = 278), that
was -log10P > 3.75 or -log10P > 4.44, was used to determine the significant or extremely significant SNP-trait
associations."   
278 individuals....

```{r}
-log10(0.05/278)
```

```{r}
thresh1=-log10(.05/67)
thresh1
thresh2=-log10(.01/67)
thresh2
```

```{r}
significant<-df2 %>% filter(log10p > thresh1) %>% filter(log10p<thresh2)
highlysignificant<-df2 %>% filter(log10p > thresh2)

nrow(highlysignificant)
```

```{r}
ggplot(df2) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=significant, aes(x=Position,y=log10p), pch=21, fill="turquoise4", alpha=0.9) +
  geom_point(data=highlysignificant, aes(x=Position,y=log10p), pch=21, fill="cyan", alpha=0.9) +
  geom_hline(yintercept=thresh1) +
  geom_hline(yintercept=thresh2, alpha=0.75, linetype="dotted") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=10, scales="free_x") +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("outputs/1200/assoc05-new-genome.pdf")
```

Adjusted p values. Removing non-sensical values, doesn't seen to matter.    

```{r}
dff<-read_tsv(file="outputs/1200/assoc05.lrt0.gz") %>%
  mutate(log10p = -log10(dchisq(LRT, df = 1)))  %>% filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

dff %>% arrange(adjp, -log10p) %>% head()                                       
```
Iso says, since the sample size is so small to sue the top sites. Also to try doAsso 2 and specify a model (lower mean counts to four or five to accomodate for less samples).

Iso would take the top SNPs and genotype them (20?) and polarize if possible, but at least genotype and make a heatmap. Major/minor. Homozygous Major, heterozygous, homozygous minor
individuals on the Y
Position on the X

Random Forest?    

Association testing, 67 individuals 75% thresh, 50.25, but, doAsso 2:
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome05.bamlist \
-yBin $HOME/ds-lh/phenos/2012-newgenome05.phenos -minMapQ 20 -minQ 20 -minInd 51 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1200/assoc05-2 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1200/assoc05-2.out 2> outputs/1200/assoc05-2.err &
```

```{r}
test1<-read_tsv(file="outputs/1200/assoc05-2.lrt0.gz") %>%
  filter(LRT != -999) %>%
  mutate(log10p = -log10(dchisq(LRT, df = 1)))  %>% 
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni")) %>%
  arrange(adjp, -log10p)

test1
```

`maccamp@farm:~/ds-lh$ gunzip -c outputs/1200/assoc05-2.lrt0.gz | sort -k 7 -n | tail`    
lg09	7338206	C	T	0.149206	67	11.798629	49/14/1     
lg18	3706822	A	T	0.348031	67	12.229991	24/34/3    
lg21	47867	C	T	0.279261	67	12.441311	26/14/0   
lg08	1030692	C	T	0.155987	67	12.640756	42/15/0    
lg22	15066999	T	C	0.492235	67	12.944254	3/31/10    
lg04	2048952	G	A	0.281251	67	12.950604	33/29/4   
lg01	12323378	C	G	0.203879	67	13.966591	28/23/0   
lg04	9935218	T	G	0.160342	67	15.772572	36/16/0   
lg02	11230311	A	C	0.293849	67	19.098647	24/17/1   
lg02	11230464	T	G	0.317309	67	19.659466	35/21/10   

Association testing, 67 individuals 75% thresh, 50.25, but, doAsso 2 and minCount = 5:
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012-newgenome05.bamlist \
-yBin $HOME/ds-lh/phenos/2012-newgenome05.phenos -minMapQ 20 -minQ 20 -minInd 51 -doAsso 2 -doPost 1 -minCount 5 -GL 1 \
-out $HOME/ds-lh/outputs/1200/assoc05-2-min -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1200/assoc05-2-min.out 2> outputs/1200/assoc05-2-min.err &
```

```{r}
test2<-read_tsv(file="outputs/1200/assoc05-2-min.lrt0.gz") %>%
  filter(LRT != -999) %>%
  mutate(log10p = -log10(dchisq(LRT, df = 1)))  %>% 
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni")) %>%
  arrange(adjp, -log10p)

test2
```


Imputation version. Maybe I should relax the -minInd flag next time (-minInd 63)     
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -GL 1 -out $HOME/ds-lh/outputs/1200/impute05 -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1 -bam $HOME/ds-lh/bamlists/2012-newgenome05.bamlist -doGlf 2  -rf outputs/1200/regions.txt  >outputs/1200/impute05.out 2>outputs/1200/impute05.err &

srun -p high -t 12:00:00 --mem=32G --nodes=2 java -Xmx15000m -jar $HOME/beagle/beagle.jar like=$HOME/ds-lh/outputs/1200/impute05.beagle.gz out=$HOME/ds-lh/outputs/1200/beagleOut05

#problem with scaffold naming using rf earlier, keeping just lgs results in 55851 sites    
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -doMaf 4 -beagle $HOME/ds-lh/outputs/1200/beagleOut05.impute05.beagle.gz.gprobs.gz -fai $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa.fai -yBin $HOME/ds-lh/phenos/2012-newgenome05.phenos -doAsso 2 -out $HOME/ds-lh/outputs/1200/impute-assoc05
```

With minInd 63    
lg12	10551688	T	C	0.317124	67	11.167774	32/24/7   
lg18	2074064	G	A	0.172016	67	11.508336	45/21/1   
lg22	15066730	G	A	0.341659	67	11.549593	22/43/1    
lg04	2048952	G	A	0.265374	67	11.800949	34/27/4   
lg08	1030692	C	T	0.158623	67	12.226135	41/17/0    
lg22	15066999	C	T	0.476485	67	12.249267	20/25/15   
lg22	15066982	C	T	0.485710	67	12.760548	19/25/19   
lg01	12323378	C	G	0.200593	67	14.367239	33/21/0   
lg04	9935218	T	G	0.159199	67	15.686034	36/16/0    
lg02	11230464	T	G	0.317326	67	19.649618	35/21/10     

Without minInd    
lg22	15066982	C	T	0.479160	67	11.869876	19/26/18   
lg04	2048952	G	A	0.267095	67	12.058967	32/27/4   
lg08	1030692	C	T	0.157656	67	12.067328	41/17/0    
lg21	47867	C	T	0.277992	67	12.334598	29/14/2    
lg18	3717798	A	G	0.327699	67	12.610455	28/29/4   
lg18	3717818	C	T	0.327604	67	12.622021	28/29/4    
lg01	12323378	C	G	0.207421	67	13.199519	28/24/0    
lg04	9935218	T	G	0.159981	67	15.746009	36/16/0    
lg02	11230311	A	C	0.314986	67	19.320573	35/20/11    
lg02	11230464	T	G	0.320277	67	19.766663	35/21/11     