---
title: "1201-local-pca"
author: "Mac Campbell"
date: "2/17/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(lostruct)
library(RColorBrewer)
library(vcfR)
```

## Local PCA

Working with 67 individuals, trying .9
```{sh, eval=FALSE}
srun -t 36:00:00 -p high --mem=32G --nodes=2 $HOME/angsd/angsd -minInd 61 -nthreads 12 \
-bam $HOME/ds-lh/bamlists/2012-newgenome05.bamlist -rf outputs/1200/regions.txt \
-out outputs/1201/localpca \
-minMaf 0.05 -minMapQ 20 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 2 -doPost 1 -postCutoff 0.9 -doPlink 2  > outputs/1201/larger.out 2> outputs/1201/larger.err &
```

Need to move locally, then:     
`plink --tped localpca.tped --tfam localpca.tfam --recodeA --out localpca-geno-numeric --allow-extra-chr --noweb`    
11166 variants and 67 people pass filters and QC.     
 
## Analysis

```{r,eval=FALSE}
snps <- read_delim("outputs/1201/localpca-geno-numeric.raw", delim = " ") %>%
  select(-IID, -SEX, -PAT, -MAT, -PHENOTYPE)
save(snps, file="outputs/1201/snps.rda")
```

Yue's analysis.

```{r}
load("outputs/1201/snps.rda")
##replace NA with the most frequent genotype
gen<-snps

#df <- snps %>% dplyr:: select(starts_with("NC_035081.1"))
#gen <- apply(df, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))

gen <- apply(gen, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))

```

```{r}
#mdsvalues<-function(chrom) {
ws=20 #window size of SNPs.The last window will be omitted as it is less than window size
mds=3 # MDS axes.
snp <- t(gen) ###transpose it so rows as SNPs, columns as individuals.
mat <- as.matrix(snp)
pcs <- eigen_windows(mat, k=2, win=ws)
pcdist <- pc_dist(pcs,npc=2)
mds <- cmdscale(pcdist, eig=TRUE, k=mds)
mds.coords <- mds$points
colnames(mds.coords) <- paste0("MDS", 1:ncol(mds.coords))
mds.coords <- mds.coords %>%
  as.data.frame() %>%
  mutate(Window=seq(1,nrow(mds.coords),1))

#%>%  mutate(Chromosome=chrom)

#}
```

Plotting.

```{r}

outl <- boxplot.stats(mds.coords$MDS1)$out
outl_ind <- which(mds.coords$MDS1 %in% c(outl))
outl_ind
outliers<-mds.coords[outl_ind,]

ggplot(mds.coords)+
  geom_point(aes(x=Window, y=MDS1), alpha=0.5) +
  geom_point(data=outliers, aes(x=Window, y=MDS1), color="blue")
  theme_bw() + 
  theme(panel.grid = element_blank())
```


```{r}

outl2 <- boxplot.stats(mds.coords$MDS2)$out
outl_ind2 <- which(mds.coords$MDS2 %in% c(outl2))
outl_ind2
outliers2<-mds.coords[outl_ind2,]

ggplot(mds.coords)+
  geom_point(aes(x=Window, y=MDS2), alpha=0.5) +
  geom_point(data=outliers2, aes(x=Window, y=MDS2), color="blue")
  theme_bw() + 
  theme(panel.grid = element_blank())
```


```{r}

outl3 <- boxplot.stats(mds.coords$MDS3)$out
outl_ind3 <- which(mds.coords$MDS3 %in% c(outl3))
outl_ind3
outliers3<-mds.coords[outl_ind3,]

ggplot(mds.coords)+
  geom_point(aes(x=Window, y=MDS3), alpha=0.5) +
  geom_point(data=outliers3, aes(x=Window, y=MDS3), color="blue")
  theme_bw() + 
  theme(panel.grid = element_blank())
```

```{r}
ggplot() +
  geom_point(data=mds.coords, aes(x=MDS1, y=MDS2)) +
  geom_point(data=outliers, aes(x=MDS1, y=MDS2), color="blue") 
```

## Split up by chrom...

```{r}
data<-read_tsv("outputs/1201/localpca.geno.gz", col_names = FALSE)

#Last col all empty
data<-data[-ncol(data)]

genos <- data %>% select(-X1,-X2)

#Want 0 to be 0/0, 1 to be 0/1 and 2 to be 1/1
genos<-genos %>% mutate_all(funs(gsub("-1","./.",.)))
genos<-genos %>% mutate_all(funs(gsub("0","0/0",.)))
genos<-genos %>% mutate_all(funs(gsub("1","0/1",.)))
genos<-genos %>% mutate_all(funs(gsub("2","1/1",.)))
#A .vcf would have some columns

cols<-data %>% select(X1,X2)

##fileformat=VCFv4.3
##fileDate=20201007@1142
##source=radiator_v.1.1.7
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT INDS

cols <- cols %>% rename('#CHROM'=X1, POS=X2) %>% mutate(ID="NA") %>% 
  mutate(REF="A",ALT="C",QUAL=".",FILTER="PASS","INFO"="NS=181", FORMAT="GT")

vcfbody<-cbind(cols,genos)

write_tsv(vcfbody, "outputs/1201/vcf.body")
```

Made a header, then:
`Macs-MacBook-Pro-2:1201 mac$ cat vcf.header vcf.body > vcf.vcf`    
`Macs-MBP-2:1201 mac$ bcftools reheader -s names.txt -o vcf-renamed.vcf vcf.vcf`
`Macs-MacBook-Pro-2:1201 mac$ bgzip vcf-renamed.vcf`     
`Macs-MacBook-Pro-2:1201 mac$ tabix vcf-renamed.vcf.gz`      
`Macs-MacBook-Pro-2:1201 mac$ cat chroms.txt  | while read line; do bcftools view -Ob -r $line vcf-renamed.vcf.gz > $line.bcf; done;`
`Macs-MacBook-Pro-2:1201 mac$ for f in *.bcf; do bcftools index $f; done;`

Need a sample info file.

"ID"	"population"
"msp_0"	"C"
"msp_1"	"C"
"msp_2"	"A"
"msp_3"	"A"

```{r}
samples<-read_tsv("bamlists/2012-newgenome05.bamlist", col_names="ID")
samples$ID<-gsub("bams/|_R1.sort.flt.bam","",samples$ID)

population<-read_tsv("phenos/2012-newgenome05.phenos", col_names="population") %>% mutate(newPop=paste0("Pop",population)) %>% select(-population) %>% rename(population=newPop)

table<-cbind(samples, population)
write.table(table, "outputs/1201/sample_info.tsv", quote = TRUE, row.names = FALSE, sep="\t")
```

all bcf/csi files in temp:    

./run_lostruct.R -i /Users/mac/github/ds-lh/outputs/1201/temp -t snp -s 20 -m 4 -I /Users/mac/github/ds-lh/outputs/1201/sample_info.tsv -j 1201   

Rscript -e 'templater::render_template("summarize_run.Rmd",output="lostruct_results/type_snp_size_20_weights_none_jobid_1201/run_summary.html",change.rootdir=TRUE)'

visualize     

```{r}
mds<-read_csv("outputs/1201/type_snp_size_20_weights_none_jobid_1201/mds_coords.csv")
#make tidy
tidymds<-mds %>% gather(MDS, Value, 3:6)
MDS1<-filter(tidymds, MDS=="MDS1") %>% rename(MDS1=MDS) %>% rename(Value1=Value)
MDS2<-filter(tidymds, MDS=="MDS2") %>% rename(MDS2=MDS) %>% rename(Value2=Value)
MDS3<-filter(tidymds, MDS=="MDS3") %>% rename(MDS3=MDS) %>% rename(Value3=Value)
MDS4<-filter(tidymds, MDS=="MDS4") %>% rename(MDS4=MDS) %>% rename(Value4=Value)
```

```{r}
ggplot(mds)+
  geom_point(aes(x=MDS1, y=MDS2), alpha=0.75) +
  theme_bw()
```

Using IQR from boxplot.
```{r}
p1<-MDS1 %>% mutate(Index=1:n())

out <- boxplot.stats(p1$Value1)$out
out_ind <- which(p1$Value1 %in% c(out))
out_ind
outliers<-p1[out_ind,]

```

```{r}
library(RColorBrewer)

#places to put labels based on index
chroms<-p1 %>% group_by(chrom) %>% mutate(Start=min(Index), Stop=max(Index)) %>% select(chrom,Start,Stop) %>% unique() %>% 
  mutate(Position=round((Start+Stop)/2))

#Make enough colors
nb.cols <- nrow(chroms)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)


#Hmm.. I should be able to include a bunch of rectangles to offset chroms
ggplot(p1) +
  geom_rect(data=chroms, aes(xmin=Start, xmax=Stop, ymin=min(p1$Value1), ymax=max(p1$Value1)), fill=mycolors, alpha=0.25) +
  geom_point(data=p1, aes(x=Index, y=Value1, color=chrom), alpha=0.75, cex=0.5) +
  geom_point(data=outliers, aes(x=Index, y=Value1), color="black", cex=0.5) +
  theme_bw() +
  scale_x_continuous(breaks=chroms$Position, labels=chroms$chrom) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, face="bold", size=6)) +
  theme(legend.title = element_blank()) +
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  scale_color_manual(values=mycolors) +
  ylab("MDS1") +
  xlab("Chromosome")

ggsave("outputs/1201/dsm-manhattan-mds1.jpg", width=10, height=4)

```

```{r}
p2<-MDS2 %>% mutate(Index=1:n())

out2 <- boxplot.stats(p2$Value2)$out
out_ind2 <- which(p2$Value2 %in% c(out2))
out_ind2
outliers2<-p2[out_ind2,]

#places to put labels based on index
chroms<-p2 %>% group_by(chrom) %>% mutate(Start=min(Index), Stop=max(Index)) %>% select(chrom,Start,Stop) %>% unique() %>% 
  mutate(Position=round((Start+Stop)/2))

#Make enough colors
nb.cols <- nrow(chroms)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)


#Hmm.. I should be able to include a bunch of rectangles to offset chroms
ggplot(p2) +
  geom_rect(data=chroms, aes(xmin=Start, xmax=Stop, ymin=min(p2$Value2), ymax=max(p2$Value2)), fill=mycolors, alpha=0.25) +
  geom_point(data=p2, aes(x=Index, y=Value2, color=chrom), alpha=0.75, cex=0.5) +
  geom_point(data=outliers2, aes(x=Index, y=Value2), color="black", cex=0.5) +
  theme_bw() +
  scale_x_continuous(breaks=chroms$Position, labels=chroms$chrom) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, face="bold", size=6)) +
  theme(legend.title = element_blank()) +
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  scale_color_manual(values=mycolors) +
  ylab("MDS2") +
  xlab("Chromosome")

ggsave("outputs/1201/dsm-manhattan-mds2.jpg", width=10, height=4)

```

We can look at ld in lg09 and lg12
`Macs-MBP-2:1201 mac$ cat chroms.txt  | while read line; do bcftools view -Ov -r $line vcf-renamed.vcf.gz > $line.vcf; done;`
plink --vcf lg09.vcf --r2 inter-chr --ld-window-r2 0.01 --out lg09.ldf --allow-extra-chr --double-id
plink --vcf lg12.vcf --r2 inter-chr --ld-window-r2 0.01 --out lg12.ldf --allow-extra-chr --double-id

```{r}
ld09<-read.delim("outputs/1201/lg09.ldf.ld",sep="",stringsAsFactors=FALSE) %>% as_tibble() %>% arrange(R2)

ggplot(ld09) +
  geom_point(aes(x=BP_A, y=BP_B, color=R2), alpha=0.5) +
  scale_color_gradient(low="khaki1", high="red")
```

```{r}
ld12<-read.delim("outputs/1201/lg12.ldf.ld",sep="",stringsAsFactors=FALSE) %>% as_tibble() %>% arrange(R2)

ggplot(ld12) +
  geom_point(aes(x=BP_A, y=BP_B, color=R2), alpha=0.5) +
  scale_color_gradient(low="khaki1", high="red")
```

We want to calculate heterozygosity over the ranges of the lg09 local pca.
```{r}
outliers2 %>% filter(chrom=="lg09")
```

12731603 start of window 25
15411529 end of window 31
17741719 end of window 34

Macs-MBP-2:1201 mac$ bcftools view -r lg09:12731603-15411529 lg09.vcf.gz > lg09-sub.vcf
Macs-MBP-2:1201 mac$ bcftools annotate --set-id '%CHROM:%POS:' lg09-sub.vcf > lg09-sub-anno.vcf

Plot PCA at this range.

```{r}
vcf<-read.vcfR(file="outputs/1201/lg09-sub-anno.vcf")
genind<-vcfR2genind(vcf)
X <- tab(genind, NA.method="mean")
pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
df<-as_tibble(pca1$li)

df <-bind_cols(df, population)

ggplot(df, aes(x=Axis1, y=Axis2)) + 
  geom_point(aes(fill=population), pch=21, alpha=0.9) +
  theme_bw() +
  theme(panel.grid = element_blank())
```

```{r}

```

Compare to whole chrom:    
bcftools annotate --set-id '%CHROM:%POS:' lg09.vcf.gz > lg09-anno.vcf    
```{r}
vcf<-read.vcfR(file="outputs/1201/lg09-anno.vcf")
genind<-vcfR2genind(vcf)
X <- tab(genind, NA.method="mean")
pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
df<-as_tibble(pca1$li)

df <-bind_cols(df, population)

ggplot(df, aes(x=Axis1, y=Axis2)) + 
  geom_point(aes(fill=population), pch=21, alpha=0.9) +
  theme_bw() +
  theme(panel.grid = element_blank())
```