---
title: "1300-new-assignments"
author: "Mac Campbell"
date: "3/29/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

## New Assignments
Courtesy of Levi et al. Let's check them out.
```{r}
newmeta<-read_csv("metadata/new_assignments.csv")
```

At a glance, I'd like to consider FWR1 vs FWM1. FWR2 seems a lot like FWM1. BWR1 is a distinctive phenotype too. Maybe I should check variation for BWR1 when I have some key loci identified.

What do our old/new ids look like?

```{r}
newmeta %>% select(new_id, old_id) %>% head()
newmeta<-newmeta %>% separate(new_id, into=c("Year","Survey","Number"), remove=FALSE)
newmeta$Number<-as.numeric(newmeta$Number)
newmeta %>% select(new_id, old_id, Number) %>% filter(new_id != old_id)
```

Hmm... Ok then.

## Find overlap with sequenced fish

Some numbers have leading zeros, some don't.
Also, we need to drop the Wakasagis.

05/05 - replacing leading zeros.
Maybe this is enough of a string to work with:Ht16-1_20
```{sh, eval=FALSE}
cat Field\ Ids.csv |  perl -pe 's/Ht(\d\d)-(\d)_20/Ht$1-0$2_20/g' > corrected-field-ids.csv
```

```{r}
#Drop the two bad samples I know about
#Ht20-30_2012_F04
#Ht20-77_2012_E10
#These two are also "funny"
# Ht19-21_2012_E03 &  2013-SKT-6396 PC-02

#seqs<-read_csv("metadata/Field Ids.csv") %>% rename(Sample_ID=`Field ID`) %>% ## Edited 05/05
seqs<-read_csv("metadata/corrected-field-ids.csv") %>% rename(Sample_ID=`Field ID`) %>%
   filter(!`Sequence File Name` %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10",
                                                      "Ht19-21_2012_E03", "Ht19-20_2012_D03")) %>%
  separate(Sample_ID, into=c("Year","Survey","Number"), remove=FALSE)

seqs$Number<-as.numeric(seqs$Number)
seqs$Sample_ID<-gsub("-","_",seqs$Sample_ID)

#Joins by Year, Survey, Number
metanew<-left_join(seqs, newmeta)
```

```{r}
hits<-metanew %>% select(`Sequence File Name`, Sample_ID, old_id, change, new_classes, Number) %>% filter(new_classes != "NA")
nrow(hits)

metanew %>% filter(new_classes != "NA") %>% group_by(Year, aggregated_new_classes) %>% summarize(Count=n())
```

We have the same number of fish (287) as I did originally, 283 due to Wakasagis or whatevers.

```{r}
metanew %>% filter(new_classes != "NA") %>% filter(Year==2013) %>%
  group_by(Year, new_classes, aggregated_new_classes) %>% summarize(Count=n())

```

Let's analyze FWM1 and FWR1.
"New" genome bamlist is of the form: bams/Ht19-70_2012_F09_R1.sort.flt.bam

```{r}
strict2013<-metanew %>% select(`Sequence File Name`, new_classes, Year) %>% filter(Year==2013) %>% 
  filter(new_classes %in% c("FWR1", "FWM1")) %>%
  mutate(File=paste0(`Sequence File Name`, "_R1.sort.flt.bam")) %>%
  mutate(Path=paste0("bams/",File))
```

Recalculating coverage on 05/05.     

`ls | grep sort.flt.bam | grep -v bai | while read line; do samtools flagstat $line | grep mapped | head -n 1 >> counts.txt; done;`
`ls | grep sort.flt.bam | grep -v bai >> counts.files.txt`


```{r}
# include coverage
files<-read_tsv("outputs/1300/counts.files.txt", col_names="File")
counts<-read_tsv("outputs/1300/counts.txt", col_names="Counts")
counts$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts$Counts, fixed = TRUE)
comb<-bind_cols(files, counts)
comb$Counts<-as.numeric(comb$Counts)
```

The below merge is incomplete when originally done, there were 30 missing samples.     

Sometime later, 05/05/2021, I figure the difference may come Sequences with 1-9 vs 01-09.   
Turns out sequences of the form Ht27-01_2014_A01_R1.fastq weren't mapped.      
Then our meta expects Ht27-1_2014_A01     

```{r}
meta2013<-left_join(strict2013, comb) 

dat<-meta2013 %>% filter(is.na(Counts)) 
write_tsv(dat, "outputs/1300/dat.tsv")

dat
```
Fixed my problem!!!!    

Samtools says that some of these should be have reads:    
samtools flagstat Ht20-14_2012_F02_R1.sort.flt.bam     
169243 + 0 in total (QC-passed reads + QC-failed reads)    

making a file for mapping
```{sh, eval=FALSE}
cut -f 4 dat.tsv | perl -pe s'/\.sort\.flt\.bam//g' > r1missing.txt
cut -f 4 dat.tsv | perl -pe s'/\.sort\.flt\.bam//g' | perl -pe s'/R1/R2/g' > r2missing.txt
paste r1missing.txt r2missing.txt > 2013-missing.txt
cat 2013-missing.txt | perl -pe 's/-(\d)_/-0$1_/g' > 2013-missing-leading0s.txt
scp -P 2022 2013-missing-leading0s.txt farm:~/ds-lh/data/
```
`(base) maccamp@farm:~/ds-lh/data$ ./doAlign.sh 2013-missing-leading0s.txt ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa`      

Now to fix the 0s in the meta and redo.


## Now to filter on coverage.
```{r}
ggplot(meta2013) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes)
```
```{r}
top2013<-meta2013 %>% top_frac(.5, Counts)
top2013 %>% group_by(new_classes) %>% summarize(Count=n())
ggplot(top2013) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes) +
  xlim(0, max(top2013$Counts))
```
```{r}
top90<-meta2013 %>% top_frac(.9, Counts)
top90 %>% group_by(new_classes) %>% summarize(Count=n())
ggplot(top90) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes) +
  xlim(0, max(top90$Counts))
```

Let's crunch the top2013 with a GWAS.   

Create new bamlist and pheno list.
```{r}
new<-top2013 %>% select(Path)
write_tsv(new, "bamlists/top2013.bamlist", col_names = FALSE)

newpheno<-top2013 %>% select(new_classes) %>% mutate(Pheno=ifelse(new_classes=="FWR1",0,1)) %>%
  select(Pheno)
write_tsv(newpheno, "phenos/top2013.phenos", col_names = FALSE)
```

Association testing, 77 individuals 75% thresh, 57.75:
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/top2013.bamlist \
-yBin $HOME/ds-lh/phenos/top2013.phenos -minMapQ 20 -minQ 20 -minInd 51 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1300/top -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1300/top.out 2> outputs/1300/top.err &
```

82307 sites.     
```{r}
df<-read_tsv(file="outputs/1300/top.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(df, -log10p) %>% top_n(20)
df2 <- filter(df, grepl("lg",Chromosome))
highest<-df2 %>% top_frac(.001,log10p) 

```

```{r}
ggplot(df2) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=highest, aes(x=Position,y=log10p), pch=21, fill="cyan") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=10, scales="free_x") +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggplot(df2 %>% filter(Chromosome=="lg02")) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=highest %>% filter(Chromosome=="lg02"), aes(x=Position,y=log10p), pch=21, fill="cyan") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  xlab("LG02") +
  ylab("-log10p")

ggsave("outputs/1300/lg02.jpeg", width=1.5, height=3)

```

Trying the alternate model.
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/top2013.bamlist \
-yBin $HOME/ds-lh/phenos/top2013.phenos -minMapQ 20 -minQ 20 -minInd 51 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1300/top2 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1300/top2.out 2> outputs/1300/top2.err &
```

82307 sites. 
```{r}
ddf<-read_tsv(file="outputs/1300/top2.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(ddf, -log10p) %>% head()

```

Chaning minInd to 90% threshold.     
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/top2013.bamlist \
-yBin $HOME/ds-lh/phenos/top2013.phenos -minMapQ 20 -minQ 20 -minInd 68 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1300/top2-90 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1300/top290.out 2> outputs/1300/top290.err &
```

73093 sites.

```{r}
ddf<-read_tsv(file="outputs/1300/top2-90.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(ddf, -log10p) %>% head()
```

At 95%    
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/top2013.bamlist \
-yBin $HOME/ds-lh/phenos/top2013.phenos -minMapQ 20 -minQ 20 -minInd 73 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1300/top2-95 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1300/top295.out 2> outputs/1300/top295.err &
```
66938 sites.   

```{r}
dddf<-read_tsv(file="outputs/1300/top2-95.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(dddf, -log10p) %>% head()
```

Maybe I can use lgs, reduce low-coverage inds some more and/or call variants with bcftools and use plink??

Using top90

```{r}
new90<-top90 %>% select(Path)
write_tsv(new90, "bamlists/top902013.bamlist", col_names = FALSE)

newpheno90<-top90 %>% select(new_classes) %>% mutate(Pheno=ifelse(new_classes=="FWR1",0,1)) %>%
  select(Pheno)
write_tsv(newpheno90, "phenos/top902013.phenos", col_names = FALSE)
```

139 inds (*.75) 104.25    
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/top902013.bamlist \
-yBin $HOME/ds-lh/phenos/top902013.phenos -minMapQ 20 -minQ 20 -minInd 105 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1300/top2902013 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/1300/top2902013.out 2> outputs/1300/top2902013.err &
````

```