---
title: "1401-combined-classes"
author: "Mac Campbell"
date: "5/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(BSDA)
library(ggmap)
```


## Combined classes!

```{r}
metanew<-read_csv("metadata/genetics-oto-intersect-05182021_JH.csv")

strict2013<-metanew %>% select(`Sequence File Name`, new_classes, Year, `Fork Length`, lat, lon, Sample_ID) %>% filter(Year==2013) %>% 
  filter(new_classes %in% c("FWR1", "FWM1", "FWM2")) %>%
  mutate(File=paste0(`Sequence File Name`, "_R1.sort.flt.bam")) %>%
  mutate(Path=paste0("bams/",File))

```

```{r}
# include coverage
files<-read_tsv("outputs/1300/counts.files.txt", col_names="File")
counts<-read_tsv("outputs/1300/counts.txt", col_names="Counts")
counts$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts$Counts, fixed = TRUE)
comb<-bind_cols(files, counts)
comb$Counts<-as.numeric(comb$Counts)

## adding three missing samples
counts2<-read_tsv("outputs/1401/toFinish.txt", col_names="Counts")
counts2$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts2$Counts, fixed = TRUE)

counts2$File<-c("Ht20-49_2012_A07_R1.sort.flt.bam","Ht20-94_2012_F12_R1.sort.flt.bam","Ht19-38_2012_F05_R1.sort.flt.bam")
counts2$Counts<-as.numeric(counts2$Counts)

comb<-bind_rows(comb,counts2)
```

Fixed later
```{sh, eval=FALSE}
samtools flagstat Ht20-49_2012_A07_R1.sort.flt.bam | grep mapped | head -n 1 >> toFinish.txt
samtools flagstat Ht20-94_2012_F12_R1.sort.flt.bam | grep mapped | head -n 1 >> toFinish.txt
samtools flagstat Ht19-38_2012_F05_R1.sort.flt.bam | grep mapped | head -n 1 >> toFinish.txt

```

```{r}
meta2013<-left_join(strict2013, comb) 

dat<-meta2013 %>% filter(is.na(Counts)) 

dat
meta2013 %>% group_by(new_classes) %>% summarize(Count=n())
```

No missing individuals


## Now to filter on coverage.
```{r}
ggplot(meta2013) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes)
```

```{r}
top2013<-meta2013 %>% top_frac(.5, Counts) %>% mutate(Phenotype=ifelse(new_classes=="FWR1","FWR","MIG"))
top2013$Pheno<-gsub("FWR","0",top2013$Phenotype)
top2013$Pheno<-gsub("MIG","1",top2013$Pheno)
top2013 %>% group_by(new_classes) %>% summarize(Count=n())
ggplot(top2013) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes) +
  xlim(0, max(top2013$Counts))

ggplot(top2013) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Phenotype) +
  xlim(0, max(top2013$Counts))
```

```{r}
nrow(top2013)
.75*nrow(top2013)

write_csv(top2013, "outputs/1401/top-2013.csv")

```

Plot these guys:

```{r}
mapbox <- c(-122.5, 37.75, -121.25, 38.5)
sfe <- get_map(location = mapbox, source = "stamen", maptype = "terrain", zoom = 9)
base<-ggmap(sfe)
```

```{r}
toplot<-top2013  %>% mutate(Phenotype=ifelse(new_classes=="FWR1","FWR","MIG")) %>%
  group_by(Phenotype, lon, lat) %>%
   summarize(Count=n()) %>% unique()

base +
  geom_point(data=toplot, aes(x=lon, y=lat, fill=Phenotype, size=Count), alpha=0.75, pch=21) +
 # geom_label_repel(data=toplot, aes(x=lon, y=lat, label=paste0(new_classes, "-", Count))) +
  facet_wrap(.~Phenotype, ncol=2) +
  xlab("") +
  ylab("") +
  theme(axis.text = element_blank()) +
  theme(axis.ticks = element_blank()) +
  scale_fill_manual(values=c("red","blue"))

ggsave("outputs/1401/combined-classes-map-2012.jpg")
```

82 individuals, to make bamlist phenotype files, using 62 inds (75% thresh)
```{r}
write_tsv(select(top2013, Path), "bamlists/1401.bamlist", col_names = FALSE)
write_tsv(select(top2013, Pheno), "phenos/1401.phenos", col_names = FALSE)
write_tsv(select(top2013, `Fork Length`), "phenos/1401.fl.phenos", col_names = FALSE)
```

No change in numbers, but change in composition from 3 inds.

Against MIG/FWR
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1401/assoc05 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05.out 2> outputs/1401/assoc05.err &
```

```{r}
df<-read_tsv(file="outputs/1401/assoc05.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(df, -log10p) %>% head()
```

What is our significance thresh?
```{r}
-log10(0.05/nrow(df))

#or


variants<-nrow(df)
#Number of expected variants
num<-10
p.T = num/variants 
prior.odds = p.T/(1-p.T) 
pwr = 1 #upper bound for power --> upper bound for alpha 
post.odds = 0.95/(1-0.95) 
alpha = prior.odds*pwr/post.odds 
paste(signif(alpha,3)) 
-log10(alpha)
```

```{r}
df2 <- filter(df, grepl("lg",Chromosome))
highest<-df2 %>% top_frac(.001,log10p) 

ggplot(df2) +
  geom_hline(yintercept = -log10(0.05/nrow(df)), col="black", alpha=0.5) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=highest, aes(x=Position,y=log10p), pch=21, fill="cyan") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=10, scales="free_x") +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("outputs/1401/gwas.jpg")
```  

And sizes:
```{r}
MIG<-top2013 %>% filter(Phenotype == "MIG")
FWR<-top2013 %>% filter(Phenotype == "FWR")

sd(MIG$`Fork Length`)
sd(FWR$`Fork Length`)

test2013<-z.test(MIG$`Fork Length`, FWR$`Fork Length`, sigma.x=4, sigma.y=3.6)
test2013
```
Migratory fish are significantly bigger.
```{r}
ggplot(top2013, aes(x = `Fork Length`, fill=Phenotype)) +                           # Draw density above histogram
  #geom_histogram(aes(y = ..density..)) +
  geom_density(alpha = 0.5) +
  facet_wrap(.~Phenotype, ncol=1) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  xlab("\nFork Length") +
  ylab("Density\n") 

ggsave("outputs/1401/top2013-fl-density.jpg")
```

  
With phenos as covariants
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yQuant $HOME/ds-lh/phenos/1401.fl.phenos -cov $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 2 -doPost 1 -GL 1 \
-out $HOME/ds-lh/outputs/1401/assoc05-fl -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-fl.out 2> outputs/1401/assoc05-fl.err &
```


```{r}
ddf<-read_tsv(file="outputs/1401/assoc05-fl.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(ddf, -log10p) %>% head()
```

Keep in mind the model is additive (could be dominant).

##Increasing Sample Sizes
```{r}
top75<-meta2013 %>% top_frac(.75, Counts) %>% mutate(Phenotype=ifelse(new_classes=="FWR1","FWR","MIG"))
top75$Pheno<-gsub("FWR","0",top75$Phenotype)
top75$Pheno<-gsub("MIG","1",top75$Pheno)
top75 %>% group_by(new_classes) %>% summarize(Count=n())
ggplot(top75) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~new_classes) +
  xlim(0, max(top75$Counts))

ggplot(top75) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Phenotype) +
  xlim(0, max(top75$Counts))
```

```{r}
nrow(top75)
nrow(top75)*0.75
write_csv(top75, "outputs/1401/top-75.csv")
```

123 individuals, to make bamlist phenotype files, using 92 inds (75% thresh)
```{r}
write_tsv(select(top75, Path), "bamlists/1401-75.bamlist", col_names = FALSE)
write_tsv(select(top75, Pheno), "phenos/1401-75.phenos", col_names = FALSE)
write_tsv(select(top75, `Fork Length`), "phenos/1401-75.fl.phenos", col_names = FALSE)
```

MIG/FWR
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401-75.bamlist \
-yBin $HOME/ds-lh/phenos/1401-75.phenos -minMapQ 20 -minQ 20 -minInd 92 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1401/assoc05-75 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-75.out 2> outputs/1401/assoc05-75.err &
```

```{r}
dddf<-read_tsv(file="outputs/1401/assoc05-75.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(bonfer = p.adjust(p, method = "bonferroni")) %>% 
  mutate(BHfdr = p.adjust(p, method = "BH")) %>%
  group_by(Chromosome) %>%
  mutate(adjpchrom = p.adjust(p, method = "bonferroni")) %>%
  mutate(BHfdrchrom = p.adjust(p, method = "BH"))

arrange(dddf, -log10p) %>% head()
```

Calculate chromosome signficance?
https://www.nature.com/articles/s41598-020-63423-8    


Prior odds way:https://www.mv.helsinki.fi/home/mjxpirin/GWAS_course/material/GWAS2.html 

```{r}

variants<-nrow(dddf)
#Number of expected variants
num<-10
p.T = num/variants 
prior.odds = p.T/(1-p.T) 
pwr = 1 #upper bound for power --> upper bound for alpha 
post.odds = 0.95/(1-0.95) 
alpha = prior.odds*pwr/post.odds 
paste(signif(alpha,3)) 
-log10(alpha)
```

Plot:

```{r}
dddf2 <- filter(dddf, grepl("lg",Chromosome))
highest2<-dddf2 %>% top_frac(.001,log10p) 

ggplot(dddf2) +
  geom_hline(yintercept = -log10(0.05/nrow(df)), col="black", alpha=0.5) +
  geom_hline(yintercept= -log10(alpha), col="black", linetype=2, alpha=0.5) +
  geom_point(aes(x=Position, y=log10p), alpha=0.5, fill="darkgrey") +
  geom_point(data=highest2, aes(x=Position,y=log10p), pch=21, fill="cyan") +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=10, scales="free_x") +
#  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("outputs/1401/gwas-123.jpg")
```
Sizes
```{r}
MIG2<-top75 %>% filter(Phenotype == "MIG")
FWR2<-top75 %>% filter(Phenotype == "FWR")

nrow(MIG2)
nrow(FWR2)

sd(MIG2$`Fork Length`)
sd(FWR2$`Fork Length`)

test75<-z.test(MIG2$`Fork Length`, FWR2$`Fork Length`, sigma.x=4, sigma.y=3.7)
test75
```

Plotting more interestingly: https://statisticsglobe.com/ggplot2-histogram-in-r-geom_histogram-function

```{r}

ggplot(top75, aes(x = `Fork Length`, fill = Phenotype)) +            # Draw two histograms in same plot
  geom_histogram(alpha = 0.5, position = "identity") +
  scale_fill_manual(values=c("red","blue"))
 
```
```{r}
ggplot(top75, aes(x = `Fork Length`, fill=Phenotype)) +                           # Draw density above histogram
  #geom_histogram(aes(y = ..density..)) +
  geom_density(alpha = 0.5) +
  facet_wrap(.~Phenotype, ncol=1) +
  theme_bw() +
  theme(panel.grid = element_blank())
```

## Length arranged by sample date
```{r}
mm<-metanew %>% select(Sample_ID, Year, SampleDate, `Fork Length`, new_classes, station, lat, lon) %>% filter(Year==2013) %>%
  filter(new_classes %in% c("FWR1", "FWM1", "FWM2")) %>%
  mutate(Phenotype=ifelse(new_classes=="FWR1","FWR","MIG")) %>%
  separate(SampleDate, c("Day","Month","Year"))
mm$Day<-as.numeric(mm$Day)
mm$Year<-as.numeric(mm$Year)
mm$Month<-factor(mm$Month, levels = c("Jan","Feb","Mar"))
```

Distribution across months:
```{r}
mm %>% group_by(Phenotype, Month) %>% summarize(Count=n()) %>% select(Phenotype, Month, Count)
```
Distribution across days:
```{r}
mm %>% group_by(Phenotype, Day, Month) %>%  summarize(Count=n()) %>% select(Phenotype, Month, Day, Count) %>% arrange(Month, Day)
nrow(mm)
mm %>% group_by(Phenotype) %>% summarize(Count=n())
```
Distribution within sampling locations. 
```{r}
temp<-mm %>% group_by(Phenotype, Month, lat, lon, station) %>% summarize(Count=n()) %>% filter(lat > 38.2)

plot<-base + geom_point(data=temp, aes(x=lon, y=lat, size=Count, color=Phenotype)) + facet_wrap(.~Phenotype)

plot
temp

mm %>%filter(station %in% c(715, 719)) %>% group_by(Phenotype, Month, station) %>% 
  mutate(Mean=mean(`Fork Length`)) %>% mutate(SampleSize=n()) %>% 
  select(Phenotype, Month, station, Mean, SampleSize) %>% unique()
```

Not a lot of overlap in sample sizes for the fish in the canal.


These are largely +/- a day.

```{r}
ggplot(mm, aes(x = `Fork Length`, fill=Phenotype)) +
  geom_density(alpha = 0.5) +
  facet_grid(Phenotype~Month) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  ylab("Density\n") +
  xlab("\nFork Length") +
  scale_fill_manual(values=c("red","blue"))

ggsave("outputs/1401/FWR-MIG-ForkLengths.jpg")
```

Test bt months:
```{r}
tester<-function(month, dataframe) {

  MIG<-dataframe %>% filter(Month== month, Phenotype == "MIG")
  FWR<-dataframe %>% filter(Month== month, Phenotype == "FWR")

  sx<-sd(MIG$`Fork Length`)
  sy<-sd(FWR$`Fork Length`)

ztest<-z.test(MIG$`Fork Length`, FWR$`Fork Length`, sigma.x=sx, sigma.y=sy)
return(ztest)

}
```

```{r}
tester("Jan", mm)
```

```{r}
tester("Feb", mm)
```

```{r}
tester("Mar", mm)
```

FL is significantly different across Months.

## Alternatives
Trying dominant and recessive models.
```{r}
arrange(df, -log10p) %>% head()
```

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 1 -model 2 -GL 1 \
-out $HOME/ds-lh/outputs/1401/assoc05-m2 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-m2.out 2> outputs/1401/assoc05-m2.err &
```

```{r}
dfm2<-read_tsv(file="outputs/1401/assoc05-m2.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(dfm2, -log10p) %>% head()
```


```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 1 -model 3 -GL 1 \
-out $HOME/ds-lh/outputs/1401/assoc05-m3 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-m3.out 2> outputs/1401/assoc05-m3.err &
```


```{r}
dfm3<-read_tsv(file="outputs/1401/assoc05-m3.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(dfm3, -log10p) %>% head()
```

That didn't do anything!!!

Trying -minCount to 2, default is 10
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 1 -GL 1 -minCount 2 \
-out $HOME/ds-lh/outputs/1401/assoc05-mincount -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-mincount.out 2> outputs/1401/assoc05-mincount.err &
```

```{r}
dfmin<-read_tsv(file="outputs/1401/assoc05-mincount.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(dfmin, -log10p) %>% head()
```
Also with -minHigh to 5, default 10 I don't know if this matters unless the model is doAsso 2

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 1 -GL 1 -minCount 2 -minHigh 5 \
-out $HOME/ds-lh/outputs/1401/assoc05-mincount-minhigh -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-mincount-minhigh.out 2> outputs/1401/assoc05-mincount-minhigh.err &
```

## Changing model -doAsso 2, 4,5,6?
Pretty sure doAsso 2 requires -doPost 1, reducing minCount to 2....
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/1401.bamlist \
-yBin $HOME/ds-lh/phenos/1401.phenos -minMapQ 20 -minQ 20 -minInd 62 -doAsso 2 -GL 1 -minCount 2 -doPost 1 \
-out $HOME/ds-lh/outputs/1401/assoc05-doasso02 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/1401/assoc05-doasso2.out 2> outputs/1401/assoc05-doasso2.err &
```

```{r}
dfdo2<-read_tsv(file="outputs/1401/assoc05-doasso02.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

arrange(dfdo2, -log10p) %>% head()
```

