---
title: "1401.2-expanded-genotypes"
author: "Mac Campbell"
date: "5/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(gplots)
library(viridis)
library(caret)
library(randomForest)
library(class)
library(gmodels)
```

## Expanding genotypes for KNN
Using `1401-75.bamlist`   
123 * .9      
[1] 110.7     

```{sh, eval=FALSE}
srun -t 36:00:00 -p high --mem=32G --nodes=2 $HOME/angsd/angsd -minInd 111 -nthreads 24 \
-bam $HOME/ds-lh/bamlists/1401-75.bamlist -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
-out $HOME/ds-lh/outputs/1401/snps-75 \
-minMaf 0.05 -minMapQ 20 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 2 -doPost 1 -postCutoff 0.9 -doPlink 2  > outputs/1401/snps-75.out 2> outputs/1401/snps-75.err &
```

Locally:
```{sh, eval=FALSE}
 plink --tped snps-75.tped --tfam snps-75.tfam --recodeA --out snps-75-geno-numeric --allow-extra-chr --noweb
```

9074 variants and 123 people pass filters and QC.

```{r, eval=FALSE}
snps <- read_delim("outputs/1401/snps-75-geno-numeric.raw", delim = " ") %>%
  select(-IID, -SEX, -PAT, -MAT, -PHENOTYPE)
save(snps, file="outputs/1401/snps-75.rda")
```

```{r}
load("outputs/1401/snps-75.rda")
top75<-read_csv("outputs/1401/top-75.csv")
```


Combine    
```{r}
phenos <- select(top75, new_classes, Phenotype)
data<-bind_cols(phenos,snps)
cols<-phenos %>% mutate(Color=ifelse(Phenotype=="MIG","blue","red")) %>% 
  mutate(ClassColor  = ifelse(new_classes=="FWR1","red",
                              ifelse(new_classes=="FWM1", "blue",
                                     ifelse(new_classes=="FWM2", "lightblue", "grey"))))
```

Find overlap between association and SNPS
```{r}
df<-read_tsv(file="outputs/1401/assoc05-75.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>% # -log10 of p>1 results in a negative value, which seems nonsensical and doesn't affect adjp much
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(adjp = p.adjust(p, method = "bonferroni"))

tops<-arrange(df, -log10p) %>% head(n=200) %>% mutate(Site=paste0(Chromosome,"_",Position))
```
have to wonder if the major is the same. Subset SNPs to tops.    
```{r}
options<-as_tibble(colnames(data)) %>% filter(!(value %in% c("Pheno","Phenotype","FID"))) %>%
  separate(value, sep="_",  into=c("Chrom","Position","Base"), remove=FALSE) %>% mutate(Site=paste0(Chrom,"_",Position)) 
options$Position <-as.numeric(options$Position)
```
This drops scaffolds.

```{r}
tops2<-inner_join(options,tops)
nrow(tops2)
```
Subset
```{r}
snpsites<-snps[colnames(snps) %in% c(tops2$value)]
```

```{r}
heatmap.2(as.matrix(snpsites), labRow="", cexRow = 0.6,
          labCol="",
          colRow = cols$ClassColor, RowSideColors = cols$ClassColor,
          trace="none",
          key=FALSE, dendrogram=c("row"), srtCol=45, margins = c(10,5), na.color=viridis(1),
          col=viridis,
          hclustfun = function(x) hclust(x,method = 'ward.D'))

```

## KNN


```{r}
tops100<-arrange(df, -log10p) %>% head(n=100) %>% mutate(Site=paste0(Chromosome,"_",Position))
options100<-as_tibble(colnames(data)) %>% filter(!(value %in% c("Pheno","Phenotype","FID"))) %>%
  separate(value, sep="_",  into=c("Chrom","Position","Base"), remove=FALSE) %>% mutate(Site=paste0(Chrom,"_",Position)) 
options100$Position <-as.numeric(options100$Position)
```

```{r}
tops2100<-inner_join(options100,tops100)
nrow(tops2100)

snpsites100<-snps[colnames(snps) %in% c(tops2100$value)]

```

```{r}
trans<-snpsites100
trans<-na.roughfix(trans)

#get 41 random samples
randos<-floor(runif(61, min=1, max=123))
smeltTrain<-trans[randos,]
train.labels<-phenos$new_classes[randos]
phenos$Phenotype[randos] %>% as_tibble() %>% group_by(value) %>% summarize(Count=n())
```


Pretty close to 50/50

Running knn
```{r}
trControl <- trainControl(method = "repeatedcv",
                          number = 100,
                          repeats = 10)

training<-smeltTrain
training$Type<-train.labels

fit <- train(Type ~ .,
             method     = "knn",
             tuneGrid   = expand.grid(k = seq(from=1, to=31, by=2)),
             trControl  = trControl,
             metric     = "Accuracy",
             data       = training)

```

```{r}
pred <- knn(train = smeltTrain, test = trans, cl = train.labels, k=fit$finalModel$k, prob = TRUE)

phenos$OptKPred<-pred
phenos$OptKProb<-round(attr(pred, "prob"),2)
CrossTable(x = phenos$new_classes, y = phenos$OptKPred, prop.chisq=FALSE)
```

How is the model?

```{r}
fit$finalModel$k
max(fit$results$Accuracy)
ggplot(fit) +
  ylim(min(fit$results$Accuracy-.1),0.95)+
  theme_classic()

ggsave("outputs/1401/training-75.jpeg")
```

```{r}
res<-knn(train=smeltTrain, test=trans, cl=train.labels, k=fit$finalModel$k)
CrossTable(x = phenos$new_classes, y = res, prop.chisq=FALSE)
```

Visualize:
```{r}
cross<-as_tibble(phenos$new_classes) %>% rename(Phenotype=value) %>% mutate(Otolith=Phenotype)
cross$Prediction<-res
summary<-cross %>% group_by(Otolith, Prediction) %>% summarize(Count=n())

ggplot(summary, aes(Otolith, Prediction)) + 
  geom_point(aes(size = Count), colour = "darkgreen") +
  theme_classic() +
  geom_text(aes(label=Count), color="white", face="bold", size=6)+
  scale_size_continuous(range=c(10,30)) +
  theme(legend.position = "none")+
  theme(axis.title = element_text(face="bold", size=14)) +
  theme(axis.text = element_text(face="bold")) +
  ggtitle("Cross Table of Phenotype and Predicted Phenotype")+
  theme(plot.title = element_text(hjust=0.5, size=16,face="bold"))
ggsave("outputs/1401/cross-table-75.jpeg")
```