---
title: "1000-2012-analysis"
author: "Mac Campbell"
date: "June 12, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

## 2012

Let's do all these things for our 2012 data 

__1__ PCA     

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Some shameless copying from Eric A.
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}
```


```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012.bamlist \
-minInd 121 -GL 1 -out $HOME/ds-lh/outputs/1000/pca \
-doGlf 2  -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-6 -minMapQ 30 -minQ 20 > outputs/1000/pca.out 2> outputs/1000/pca.err &

#Generate Covariance Matrix
python $HOME/pcangsd/pcangsd.py -beagle $HOME/ds-lh/outputs/1000/pca.beagle.gz -admix -o $HOME/ds-lh/outputs/1000/pca
```


```{r}
library(tidyverse)
pca_meta<-read_csv("metadata/metadata-2012.csv") %>% rename(sample=`Sequence File Name`) 
cov<-read_delim("outputs/1000/pca.cov", delim=" ", col_names=FALSE) %>% as.matrix()

pca <- covar2pcs(pca_meta$sample, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pca_meta, by = "sample")

#drop outliers, deleted from bamlist and rerun
#Ht20-30_2012_F04
#Ht20-77_2012_E10
pp_meta<-pp_meta %>% filter(!(sample %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10")))
ggplot(pp_meta, aes(x = val_x, y = val_y, fill=Phenotype), alpha=0.75) +
  geom_point(pch = 21, size = 2) +
  scale_fill_discrete(na.value = "white") + 
  scale_fill_manual(values = c("red","blue")) +
  facet_grid(PCy ~ PCx, scales = "free")+
  theme_bw()

ggsave("outputs/1000/6-6-pca.pdf", width=14, height=12)



```
PCA from adegenet
```{r}

load("outputs/1000/recode.genind")
genind@pop=as.factor(pca_meta$Phenotype)

gen<-genind
X <- tab(gen, NA.method="mean")

pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
temp <- as.integer(pop(gen))
myCol <- transp(c("red","blue"),.7)[temp]
myPch <- c(15,17)[temp]
plot(pca1$li, col=myCol, cex=1, pch=myPch)

pcdf<-as_tibble(pca1$l1) %>% rename(PC1=RS1, PC2=RS2)
pcdf$Phenotype<-pop(gen)

eigens<-as_tibble(pca1$eig) %>% rename(Eigen=value) %>% mutate(Porportion=Eigen/sum(Eigen)*100)

ggplot(pcdf)+geom_point(aes(x=PC1, y=PC2, fill=Phenotype), alpha=0.75, size=2, pch=21)+
  scale_fill_manual(values = c("red","blue")) +
  xlab(paste0("PC1 ", round(eigens$Porportion[1],2),"%")) +
  ylab(paste0("PC2 ", round(eigens$Porportion[2],2),"%")) +
  theme_bw()+
  theme(panel.grid=element_blank()) +
  theme(axis.title = element_text(face="bold")) +
  theme(legend.title = element_text(face="bold"))

ggsave("outputs/1000/pc12.pdf")
```


__2__ DAPC     
Need a plink file.
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -b $HOME/ds-lh/bamlists/2012.bamlist -minInd 121 \
-out $HOME/ds-lh/outputs/1000/plink \
-minMaf 0.05 -minMapQ 30 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 \
-SNP_pval 1e-6 -doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2 > outputs/1000/plink.out 2> outputs/1000/plink.err &
```

```{sh, eval=FALSE}
cat plink.tped | perl -pe 's/^(\d+)\s\d+_/contig_$1 contig_$1_/' > temp.tped
cp plink.tfam temp.tfam
plink --tped temp.tped --tfam temp.tfam  --out binary --recode --allow-extra-chr --noweb
plink --ped binary.ped --map binary.map --recode vcf --allow-extra-chr -out recode
```

```{r, eval=FALSE}
library(adegenet)
library(vcfR)
vcf<-read.vcfR(file="outputs/1000/recode.vcf")
genind<-vcfR2genind(vcf)
save(genind, file="outputs/1000/recode.genind")
```
4848 variants
```{r}
library(adegenet)
load("outputs/1000/recode.genind")
genind@pop=as.factor(pca_meta$Phenotype)
dapc<-dapc(genind, n.pca=115, n.da=1)

#Red is FWR, Blue is MIG.    
scatter(dapc, col=c("red","blue"))
pdf("outputs/1000/dapc.pdf", width=7, height=5)
scatter(dapc, col=c("red","blue"))
dev.off()

```

loading plot
```{r}
contrib <- loadingplot(dapc$var.contr, axis=1,
thres=.0017, lab.jitter=0)
pdf("outputs/1000/loadingplot.pdf")
contrib <- loadingplot(dapc$var.contr, axis=1,
thres=.0017, lab.jitter=0)
dev.off()

vals<-contrib[contrib$var.values>.0017]

tab(genind2genpop(genind[loc=c("contig_196_382662")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_393_62098")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_116045_4660")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_116393_11807")]),freq=TRUE)

```



```{r}
post<-as_tibble(dapc$posterior)
post$Phenotype<-dapc$grp

ggplot((post %>% filter(Phenotype == "FWR")), aes(x=FWR))+
  geom_histogram(color="darkred", fill="red")+
  ggtitle("Posterior Probability of FWR Assigned to FWR")+
  xlim(-.01,1.01)+
  ylab("Count") +
  theme_classic() +
  theme(plot.title = element_text(hjust=0.5))+
  theme(axis.text = element_text(face="bold",size=12))+
  theme(axis.title = element_text(face="bold", size=14))+
  theme(plot.title = element_text(face="bold", size=16))

ggsave("outputs/1000/postFWR.pdf")

ggplot((post %>% filter(Phenotype == "MIG")), aes(x=MIG))+
  geom_histogram(color="darkblue", fill="blue")+
  ggtitle("Posterior Probability of MIG Assigned to MIG")+
  xlim(-.01,1.01)+
  ylab("Count") +
  theme_classic() +
  theme(plot.title = element_text(hjust=0.5))+
  theme(axis.text = element_text(face="bold",size=12))+
  theme(axis.title = element_text(face="bold", size=14))+
  theme(plot.title = element_text(face="bold", size=16))

ggsave("outputs/1000/postMIG.pdf")

```



__2__ Association testing    
```{sh,eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012.bamlist \
-yBin $HOME/ds-lh/phenos/2012.phenos -minMapQ 30 -minQ 20 -minInd 121 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1000/assoc -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/ds-lh/genome/Hypomesus-transpacificus_10X_F_A.pseudohap2.1.fasta > outputs/1000/assoc.out 2> outputs/1000/assoc.err &
```


I can also try imputation.    
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -GL 1 -out $HOME/ds-lh/outputs/1000/input -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1  -bam $HOME/ds-lh/bamlists/2012.bamlist -doGlf 2 >outputs/1000/impute1.out 2>outputs/1000/impute1.err &


#Using version 3.3.2 for compatibility with ANGSD
srun -p high -t 12:00:00 --mem=32G --nodes=2 java -Xmx15000m -jar $HOME/beagle/beagle.jar like=$HOME/ds-lh/outputs/1000/input.beagle.gz out=$HOME/ds-lh/outputs/1000/beagleOut

#Notes Warning: All data sets have > 0.07 missing alleles for the markers in the markers file.
#         To obtain the most accurate imputation, at least one input Beagle file should be genotyped
#         for all markers in the markers file and should have < 0.07 missing alleles.
# Can consider rerunning beagle likelihood -minInd 128 to get a 95% missing threshold

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -doMaf 4 -beagle $HOME/ds-lh/outputs/1000/beagleOut.input.beagle.gz.gprobs.gz -fai $HOME/genomes/hypomesus/Hypomesus-transpacificus_10X_F_A.pseudohap2.1.fasta.fai  -yBin $HOME/ds-lh/phenos/2012.phenos -doAsso 2 -out $HOME/ds-lh/outputs/1000/impute-assoc


```

How did the imputation work out?
```{r}
df<-read_tsv(file="outputs/1000/impute-assoc.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf")

arrange(df, -log10p) %>% top_n(20)
df2 <- df %>% filter(Chromosome %in% c("32","58","62","103","224")) 
ggplot(df2) +
  geom_point(aes(x=Position, y=log10p, color=log10p)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  facet_wrap(.~Chromosome, ncol=5, scales="free_x") +
  scale_color_gradient(low="darkgrey", high="cyan") +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

grabbed these:
`>58:5177700-5178300
CTGGATTCACAGGAAACAACCTAAATTGGACACACACACACACACCTGTATCTGGGACAT
CTGTTGGCCTCATGATCCTACGAATCTGTTCCATGGACTGCAGGTCAGGTGACAGGCCTG
CAGGTTAGGAGAATGCATTGAGCATTGATAGAATATCGGGGACAACAGGAAGACAGGACA
GAGAGCGAGCAGAAAAGAGACGCCCCCCCTCCCGTCTATCCTCTGACCTCCATGGCAGCA
GAAGATCTTCTCGTCGATGATGGCAGCTATGGGGAGACAGTTGAAGCAGTCCGTGAAGGT
CTTCCACAGCTTGATGTTGAACCTGCGCTTACCTGGTGTGTGTGTGTGTGGGGGGGGGGG
GACACAGAGAAAAGCCGTAAATGTCACCGTGTGTGTGTATGTGTACACAGTGACTGGGAG
TGAGAGAATGTACTGTACGTGTGCGTTCACATGCTCGCTCACACTCATCGTAGAAGCCGT
AGATGCGGTTGATGGACGCACACTCGTGGTTTCCCCGCAGCAGGAAGAAGTTCTCCGGGT
ACTTGATCTTGTAGGCCAGCAGGAGACAGATGGTTTCCAGGGACTGCTTCCCTCGGTCCA
C`

As and ORF:
>lcl|ORF4
MSVSEHVNAHVQYILSLPVTVYTYTHTVTFTAFLCVPPPPHTHTPGKRRF
NIKLWKTFTDCFNCLPIAAIIDEKIFCCHGGQRIDGRGGVSFLLALCPVF
LLSPIFYQCSMHSPNLQACHLTCSPWNRFVGS

So this is a protein: 
"Serine/threonine-protein phosphatase PP1-beta catalytic subunit; Short=PP-1B"    

Aligns to "Salarias fasciatus chromosome 19, fSalaFa1.1, whole genome shotgun sequence, 11305495-11305846"    
`join(<1..139,246..>350)    
                     /gene="LOC115406247"    
                     /product="serine/threonine-protein phosphatase PP1-beta    
                     catalytic subunit-like`   

#Now to impute for the 202 fish data set
```{sh, eval=FALSE}

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -GL 1 -out $HOME/ds-lh/outputs/1000/input202 -doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1 -minInd 192 -bam $HOME/ds-lh/bamlists/202.bamlist -doGlf 2 >outputs/1000/impute202.out 2>outputs/1000/impute202.err &

srun -p high -t 12:00:00 --mem=32G --nodes=2 java -Xmx15000m -jar $HOME/beagle/beagle.jar like=$HOME/ds-lh/outputs/1000/input202.beagle.gz out=$HOME/ds-lh/outputs/1000/beagleOut202

srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -doMaf 4 -beagle $HOME/ds-lh/outputs/1000/beagleOut202.input.beagle.gz.gprobs.gz -fai $HOME/genomes/hypomesus/Hypomesus-transpacificus_10X_F_A.pseudohap2.1.fasta.fai  -yBin $HOME/ds-lh/phenos/202.phenos -doAsso 2 -out $HOME/ds-lh/outputs/1000/impute-assoc-202

```


```{r, warning=FALSE, message=FALSE}
data<-read_tsv(file="outputs/1000/assoc.lrt0.gz")

#Converting to log10p
data <- data %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) 

# To be explicit, our p - values convert like so
# -log10(0.01) = 2
# -log10(0.001) = 3
# -log10(0.0001) = 4

##Example manhattan plot
sub<-data %>% filter(Chromosome %in% c(32,52,255)) %>% filter(log10p > 0)
sub$Chromosome<-gsub("32","Chrom 1",sub$Chromosome)
sub$Chromosome<-gsub("52","Chrom 2",sub$Chromosome)
sub$Chromosome<-gsub("255","Chrom 3", sub$Chromosome)

sub$Chromosome<-factor(sub$Chromosome)
fake<-read_tsv("outputs/1000/fake.tsv")

sub<-bind_rows(sub,fake)
ggplot(sub) +
  geom_point(aes(x=Position, y=log10p, color=Chromosome), alpha=0.95, cex=1.5) +
  facet_wrap(~ Chromosome, ncol=4, scales="free_x")+
  ylim(0,15) +
  geom_hline(yintercept = -log10(5E-8)) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  theme(legend.position = "none")+
  ylab("-log10(p)")
ggsave("outputs/1000/manhattan.pdf",width=3, height=5)

#Check for most significant values
data %>% filter(log10p > 3) %>% arrange(-log10p)

chroms<-data %>% top_n(log10p, n=30) %>% group_by(Chromosome) %>% summarize(Count=n()) %>% arrange(-Count)

#read_tsv(file="outputs/300/202.lrt0.gz") %>% arrange(-LRT) %>% mutate(log10p=-log10(dchisq(LRT, df = 1))) %>% filter(Chromosome==58)
data$ChromosomeFact<-factor(data$Chromosome)
ggplot(data%>%filter(Chromosome %in% c(54,62,87,113,117,332,117438) & log10p>=0.5))+
  geom_point(aes(x=Position, y=log10p, color=ChromosomeFact), alpha=0.95, cex=1.5)+
  #geom_hline(yintercept =4, size=0.5, alpha=0.5)+
  facet_wrap(~ Chromosome, ncol=7, scales="free_x")+
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1))+
  theme(panel.grid = element_blank())+
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  ylab("-log10(p) ANGSD") +
  theme(axis.title=element_text(face="bold",size=14))+
  theme(legend.position = "none")
ggsave("outputs/1000/gwas-realdata.pdf")
```

###Plink GWAS
```{sh, eval=FALSE}
vcftools --vcf outputs/1000/recode.vcf --plink --out outputs/1000/plink-recode

plink --file outputs/1000/plink-recode --allow-no-sex --allow-extra-chr --make-bed --noweb --out outputs/1000/plink-recode.binary

#Alternate alleles
cat outputs/1000/recode.vcf | awk 'BEGIN{FS="\t";OFS="\t";}/#/{next;}{{if($3==".")$3=$1":"$2;}print $3,$5;}'  > outputs/1000/alt_alleles

#List of phenotypes
cat phenos/2012.phenos | perl -pe 's/0/FWR/g' | perl -pe 's/1/MIG/g' | awk '{print NR"_1 "NR"_1 " $1}' > outputs/1000/pheno.txt

plink --bfile outputs/1000/plink-recode.binary --make-pheno outputs/1000/pheno.txt MIG --assoc --reference-allele outputs/1000/alt_alleles --allow-no-sex --allow-extra-chr --adjust --noweb --out outputs/1000/plink-gwas

```


```{r}
library(ggrepel)
plink<-read_table(file = "outputs/1000/plink-gwas.assoc.adjusted")

#Getting the top 1%
top1<-plink %>% top_frac(-.01, GC)

data2<-read_tsv(file="outputs/200/216.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>% mutate(SNP = paste0("contig_",Chromosome,"_",Position))
top2<-data2 %>% top_frac(.01, log10p)

comb<-inner_join(top1, top2)
comb
top5<-head(comb,5)
total<-inner_join(plink,data2)
ggplot(total, aes(x=log10p, y=-log10(GC)))+geom_point(color="darkgrey",alpha=0.75)+
  geom_point(data=comb, aes(x=log10p, y=-log10(GC)), color="red", alpha=0.75, cex=1.5)+
  geom_text_repel(data=top5, aes(x=log10p,y=-log10(GC),label=paste0(Chromosome,"-",Position)))+
  theme_bw()+
  theme(panel.grid = element_blank()) +
  ylab("-log10(p) PLINK") +
  xlab("-log10(p) ANGSD") +
  theme(axis.title = element_text(face="bold"))

ggsave("outputs/1000/gwas-combined.pdf")
#Where are most of these found?
comb %>% group_by(Chromosome) %>% summarize(Count=n()) %>% arrange(-Count)
write.csv(comb, file="outputs/1000/combined-gwas.csv")

```

__3__ KNN prediction    

Need numeric .geno.gz

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -b $HOME/ds-lh/bamlists/2012.bamlist \
-minInd 121 -out $HOME/ds-lh/outputs/1000/2012-genos \
-minMaf 0.05 -minMapQ 30 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 \
-SNP_pval 1e-6 -doGeno 2 -doPost 1 -postCutoff 0.95 > outputs/1000/2012-genos.out 2> outputs/1000/2012-genos.err &
```
