---
title: "1000-2012-analysis"
author: "Mac Campbell"
date: "June 12, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

## 2012

Let's do all these things for our 2012 data 

__1__ PCA     

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Some shameless copying from Eric A.
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}
```


```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012.bamlist \
-minInd 123 -GL 1 -out $HOME/ds-lh/outputs/1000/pca \
-doGlf 2  -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-6 -minMapQ 30 -minQ 20 > outputs/1000/pca.out 2> outputs/1000/pca.err &

#Generate Covariance Matrix
python $HOME/pcangsd/pcangsd.py -beagle $HOME/ds-lh/outputs/1000/pca.beagle.gz -admix -o $HOME/ds-lh/outputs/1000/pca
```


```{r}
library(tidyverse)
pca_meta<-read_csv("metadata/metadata-2012.csv") %>% rename(sample=`Sequence File Name`) 
cov<-read_delim("outputs/1000/pca.cov", delim=" ", col_names=FALSE) %>% as.matrix()

pca <- covar2pcs(meta$`Sequence File Name`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pca_meta, by = "sample")

#drop outliers, deleted from bamlist and rerun
#Ht20-30_2012_F04
#Ht20-77_2012_E10
pp_meta<-pp_meta %>% filter(!(sample %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10")))
ggplot(pp_meta, aes(x = val_x, y = val_y, fill=Phenotype)) +
  geom_point(pch = 21, size = 2) +
  scale_fill_discrete(na.value = "white") + 
  facet_grid(PCy ~ PCx, scales = "free")+
  theme_bw()

ggsave("outputs/1000/6-6-pca.pdf", width=14, height=12)

```
__2__ DAPC     
Need a plink file.
```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -b $HOME/ds-lh/bamlists/2012.bamlist -minInd 123 \
-out $HOME/ds-lh/outputs/1000/plink \
-minMaf 0.05 -minMapQ 30 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 \
-SNP_pval 1e-6 -doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2 > outputs/1000/plink.out 2> outputs/1000/plink.err &
```

__2__ Association testing    
```{sh,eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/ds-lh/bamlists/2012.bamlist \
-yBin $HOME/ds-lh/phenos/2012.phenos -minMapQ 30 -minQ 20 -minInd 123 -doAsso 1 -GL 1 \
-out $HOME/ds-lh/outputs/1000/assoc -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/ds-lh/genome/Hypomesus-transpacificus_10X_F_A.pseudohap2.1.fasta > outputs/1000/assoc.out 2> outputs/1000/assoc.err &
```
__3__ KNN prediction    

Need numeric .geno.gz

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24 -b $HOME/ds-lh/bamlists/2012.bamlist \
-minInd 123 -out $HOME/ds-lh/outputs/1000/2012-genos \
-minMaf 0.05 -minMapQ 30 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 \
-SNP_pval 1e-6 -doGeno 2 -doPost 1 -postCutoff 0.95 > outputs/1000/2012-genos.out 2> outputs/1000/2012-genos.err &
```
