---
title: "300-RF"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# DAPC With 216 samples.


```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(adegenet)
load("outputs/200/recode.genind")
load("outputs/200/filtered.rda")
genind@pop=as.factor(filtered$Phenotype)
drop<-c("89_1","90_1")
gen <- genind[!row.names(genind@tab) %in% drop]
nInd(gen)

#PCA
X <- tab(gen, NA.method="mean")

pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
temp <- as.integer(pop(gen))
myCol <- transp(c("red","blue"),.7)[temp]
myPch <- c(15,17)[temp]
plot(pca1$li, col=myCol, cex=1, pch=myPch)

pcdf<-as_tibble(pca1$l1) %>% rename(PC1=RS1, PC2=RS2)
pcdf$Phenotype<-pop(gen)

ggplot(pcdf)+geom_point(aes(x=PC1, y=PC2, color=Phenotype), alpha=0.75)+
  scale_color_manual(values=c("red","blue"))+
  theme_bw()+
  theme(panel.grid=element_blank())
  
#DAPC
dapc<-dapc(gen, n.pca=175, n.da=1)
#Red is FWR, Blue is MIG.    
scatter(dapc, col=c("red","blue"))
contrib <- loadingplot(dapc$var.contr, axis=1,
thres=.0017, lab.jitter=0)
```

Wow! Loadings are bigger than ever.

```{r, warning=FALSE, message=FALSE}
tab(genind2genpop(genind[loc=c("contig_103_481573")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_571477")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_2863421")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_105394_1469")]),freq=TRUE)
```

Some version of an assignment plot.    
```{r, warning=FALSE, message=FALSE}
post<-as_tibble(dapc$posterior)
post$Phenotype<-dapc$grp

ggplot((post %>% filter(Phenotype == "FWR")), aes(x=FWR))+
  geom_histogram(color="darkred", fill="red")+
  ggtitle("Posterior Probability of FWR Assigned to FWR")+
  theme_bw()+
  theme(panel.grid=element_blank())
  
ggplot((post %>% filter(Phenotype == "MIG")), aes(x=MIG))+
  geom_histogram(color="darkblue", fill="blue")+
  ggtitle("Posterior Probability of MIG Assigned to MIG")+
  theme_bw()+
  theme(panel.grid=element_blank())

```

# Load in pruned data

```{r, warning=FALSE, message=FALSE, eval=FALSE}
library(vcfR)
vcf<-read.vcfR(file="outputs/200/recode.prune.vcf")
genprune<-vcfR2genind(vcf)
genprune@pop=as.factor(filtered$Phenotype)
genprune <- genprune[!row.names(genind@tab) %in% drop]
save(genprune, file="outputs/300/pruned.genind")
dapc2<-dapc(genprune, n.pca=175, n.da=1)
scatter(dapc2, col=c("red","blue"))
```

There are quite a few of those FWR fish at the tail end overlapping with MIG fish.
```{r, warning=FALSE, message=FALSE}
load("outputs/300/pruned.genind")
dapc2<-dapc(genprune, n.pca=175, n.da=1)
scatter(dapc2, col=c("red","blue"))

post2<-as_tibble(dapc2$posterior)
post2$Phenotype<-dapc2$grp
post2$Ind<-indNames(genprune)

FWR<-post2 %>% filter(Phenotype=="FWR" & FWR > 0.95)
MIG<-post2 %>% filter(Phenotype=="MIG" & MIG > 0.95)

#Pulling out the posterior probabilities for group assignments
temp1<-post2 %>% select (FWR, Phenotype) %>% filter(Phenotype=="FWR",FWR < 0.95)
temp2<-post2 %>% select (MIG, Phenotype) %>% filter(Phenotype=="MIG",MIG < 0.95)
# pps of ~ 0.75 and 0.50
ggplot(temp1)+geom_histogram(aes(x=FWR))
ggplot(temp2)+geom_histogram(aes(x=MIG))
combined<-rbind(FWR, MIG) # now 202 individuals

sub<-genprune[row.names(genprune@tab) %in% combined$Ind]

dapc3<-dapc(sub, n.pca=175, n.da=1)

scatter(dapc3, col=c("red","blue"))

```

OOhhh, what are loadings looking like now?

```{r, warning=FALSE, message=FALSE}
contrib <- loadingplot(dapc3$var.contr, axis=1,
thres=.0017, lab.jitter=0)

tab(genind2genpop(genind[loc=c("contig_103_481573")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_571477")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_2863421")]),freq=TRUE)

```

I will check to see if those 14 individuals had an effect on association tests. Creating a new bamlist and pheno list.

```{r, warning=FALSE, message=FALSE}

filtered$ID<-indNames(genind)

filtered202<-filtered %>% filter(ID %in% indNames(sub))

bams<-select(filtered202, `Sequence File Name`) %>% mutate(Path=paste0("bams/",`Sequence File Name`,".sort-n.fixmate-m.sort.markdup-r.bam")) %>% select(Path)
write_tsv(bams, path = "bamlists/202.bamlist", col_names = FALSE)

phenos<-select(filtered202,Phenotype)
phenos$Phenotype<-gsub("FWR",0,phenos$Phenotype)
phenos$Phenotype<-gsub("MIG",1,phenos$Phenotype)
write_tsv(phenos, path = "phenos/202.phenos", col_names=FALSE)

```


And to check it out... No real difference in outcome.    

```{r, warning=FALSE, message=FALSE}

asso<-read_tsv(file="outputs/300/202.lrt0.gz")

#Converting to log10p, and removing p - values > 0.01
asso <- asso %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p > 2)

asso %>% filter(log10p > 3.5) %>% arrange(desc(log10p))

```

## What are those 16 funny fish about
2 at extremes of PCA. Are they real? Don't know.     
14 that seemed like hybrids/miss-assigned, of those, two may be some sort of mislabelling. I don't know.     

```{r, warning=FALSE, message=FALSE}
table<-as_tibble(dapc$posterior)
table$Samples<-row.names(dapc$posterior)

#        FWR   MIG Samples
#      <dbl> <dbl> <chr>  
# 1 1.93e- 3 0.998 13_1   
# bams/Ht19-31_2012_G04.sort-n.fixmate-m.sort.markdup-r.bam

#Also 132_1
#bams/Ht20-30_2012_F04.sort-n.fixmate-m.sort.markdup-r.bam

samples<-read_tsv("outputs/100/meta2.rda")
samples %>% filter(`Sequence File Name` %in% c("Ht19-31_2012_G04","Ht20-30_2012_F04"))
```

## One last burning question, what if I make an arbitrary division within a group?
I'm going to take 60 FWR and split in half
```{r, warning=FALSE, message=FALSE}
pop1<-rep("pop1", 30)
pop2<-rep("pop2", 30)
fwr<-sub[1:60]
fwr@pop<-as.factor(c(pop1,pop2))
fw<-dapc(fwr, n.pca=60, n.da=1)
scatter(fw)

loadingplot(fw$var.contr, axis=1,
thres=.0017, lab.jitter=0)

#Or try to divide FWR
##Search for signal in the mig
migr<-sub[62:202]

grps<-find.clusters(migr, n.pca=110, n.clust = 2)
scatter(migdapc)
migr2<-migr
migr2@pop=grps$grp
migdapc<-dapc(migr, pop=grps$grp, n.pca=135, n.da=1)
loadingplot(migdapc$var.contr, axis=1,
thres=.0017, lab.jitter=0)

#And some allele frequencies
tab(genind2genpop(migr2[loc=c("contig_54_712342")]),freq=TRUE)
tab(genind2genpop(migr2[loc=c("contig_92_861840")]),freq=TRUE)
tab(genind2genpop(migr2[loc=c("contig_92_885894")]),freq=TRUE)

```

60 sample GWAS. I'll take these same 60 samples and see if I get a result from the GWAs.



# And now to prep data for RF

We probabaly want our pruned 202 data set reformatted for Random Forest
```{r, warning=FALSE, message=FALSE}
library(randomForest)
temp<-read.csv("~/random-forest/data_classification_RF_tutorial.csv", row.names=1)
genos<-read_tsv("outputs/300/202.geno.gz", col_names = FALSE)

#Removing last column and transposing
genos<-genos %>% select(-X205)
trans<-as_tibble(t(genos))
trans<-trans[-1:-2,]
names<-read_tsv("bamlists/202.bamlist",col_names=FALSE) %>% mutate(Names=gsub("bams/","",X1))
names <- names %>% mutate(Sample=gsub(".sort-n.fixmate-m.sort.markdup-r.bam","",Names)) %>% select(Sample)

phenos <- read_tsv("phenos/202.phenos", col_names=FALSE) %>% rename(Pop=X1) %>% mutate(Phenotype=Pop)

meta<-as_tibble(cbind(names,phenos))

input<-as_tibble(cbind(meta,trans))

#For testing, let's use the first 1000 loci to match tutorial
sub<-input[1:1002]

```

Now we have the data in the proper input format, but, I do wonder about the missing (-1 data). RF should be able to handle that, there is the rfImpute fucnction or specifying na.action, am going the rough.fix route.

```{r, warning=FALSE, message=FALSE}
#Using df for generalization
df<-sub
df[df < 0] <- NA

length(which(input$Phenotype==0)) # 61 here today
length(which(input$Phenotype==1)) # 141 here today
sample_size <- c(61,61)

results_optimization <- matrix(data=NA , nrow = 0, ncol = 3)
for (i in seq(from = 100, to = 1000 , by = 100)){  # values of ntree
  print(i)
  for (j in c(32,64,100,200,333,1000)){    #values of mtry based on 1000 total loci
    rf_ij <- randomForest(x = na.roughfix(df[,3:1002]), y = as.factor(df$Phenotype), importance=TRUE ,proximity=TRUE, ntree=i, mtry=j, strata=as.factor(df$Phenotype), sampsize=sample_size)
    results_optimization <- rbind(results_optimization, c(i,j,tail(rf_ij$err.rate,1)[1]))
  }
}



```

```{r, warning=FALSE, message=FALE}
results_optimization<-as.data.frame(results_optimization)
colnames(results_optimization)<-c("ntree", "mtry","OOB_ER")

# Now plot results to see if there's a plateau
plot(results_optimization$ntree[results_optimization$mtry == 32],results_optimization$OOB_ER[results_optimization$mtry == 32], type="l", col="black", xlab="ntree",ylab="OOB-ER",ylim=c(0,1))
lines(results_optimization$ntree[results_optimization$mtry == 64],results_optimization$OOB_ER[results_optimization$mtry == 64], col="blue")
lines(results_optimization$ntree[results_optimization$mtry == 100],results_optimization$OOB_ER[results_optimization$mtry == 100], col="green")
lines(results_optimization$ntree[results_optimization$mtry == 200],results_optimization$OOB_ER[results_optimization$mtry == 200], col="purple")
lines(results_optimization$ntree[results_optimization$mtry == 333],results_optimization$OOB_ER[results_optimization$mtry == 333], col="orange")
lines(results_optimization$ntree[results_optimization$mtry == 1000],results_optimization$OOB_ER[results_optimization$mtry == 1000], col="red")



```


Ok! I think I'm ready to set up a series of scripts for the actual RF analsyis.