---
title: "500-first-summary"
author: "Mac Campbell"
date: "1/23/2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# First Analyses of Combined Otolith and Genetic Data
This is meant to show some key results of what I have been doing with the two data types. I will avoid getting into a lot of details about the methodology.

## Surviving Inputs
Initially I filtered the data sets for two things. (1) Overlap between otolith and genetic data, and (2) sequencing depth of individuals with genetic data. I omitted the bottom 15% of samples in terms of coverage. This results in 72 FWR and 146 MIG fish with sequence data.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6}
library(tidyverse)
library(gridExtra)
samples<-read_tsv("outputs/100/meta2.rda")
bams<-read_tsv("bamlists/full.bamlist",col_names = FALSE) %>% mutate(`Sequence File Name`=gsub(".sort-n.fixmate-m.sort.markdup-r.bam","",X1)) %>% mutate(Bams=X1)
bams$`Sequence File Name`<-gsub("bams/","",bams$`Sequence File Name`)

coverage<-read_tsv("outputs/200/full.coverage", col_names = FALSE)

combined<-inner_join(bams,samples)

combined$Coverage<-coverage$X1

p1<-ggplot(combined)+
  geom_histogram(aes(x=Coverage))+
  facet_grid(.~Phenotype)+
  theme_bw()+
  theme(panel.grid = element_blank())+
  xlim(0,20)+
  ggtitle("Coverage of Samples with Otolith Data")+
  theme(plot.title = element_text(hjust = 0.5))


filtered<- combined %>% top_frac(.85, Coverage)

p2<-ggplot(filtered)+
  geom_histogram(aes(x=Coverage))+
  facet_grid(.~Phenotype)+
  theme_bw()+
  theme(panel.grid = element_blank())+
  xlim(0,20)+
  ggtitle("Average Coverage of Top 85% of Samples with Otolith Data")+
  theme(plot.title = element_text(hjust = 0.5))

grid.arrange(p1, p2, nrow = 2)

filtered %>% select(Phenotype) %>% group_by(Phenotype) %>% summarize(Count=n())
```

## Basic Genetic Variation
As a first pass, I conducted a principal component analysis of the 218 fish and 4,719 loci. This showed two samples that were very distinctive classified as MIG, perhaps misidentified fish (not shown). Plotting of the remaining 216 fish follows our expectations of no indications of genetic structuring in this species. Overall summary statistics do not appear to be different between phenotypes (output from the poppr function shown https://www.rdocumentation.org/packages/poppr/versions/2.8.3/topics/poppr).    

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6}
library(adegenet)
library(poppr)
load("outputs/200/recode.genind")
load("outputs/200/filtered.rda")
genind@pop=as.factor(filtered$Phenotype)
drop<-c("89_1","90_1")
gen <- genind[!row.names(genind@tab) %in% drop]
#nInd(gen)

#PCA
X <- tab(gen, NA.method="mean")

pca1 <- dudi.pca(X,scannf=FALSE,scale=FALSE)
temp <- as.integer(pop(gen))
myCol <- transp(c("red","blue"),.7)[temp]
myPch <- c(15,17)[temp]
#plot(pca1$li, col=myCol, cex=1, pch=myPch)

pcdf<-as_tibble(pca1$l1) %>% rename(PC1=RS1, PC2=RS2)
pcdf$Phenotype<-pop(gen)

ggplot(pcdf)+geom_point(aes(x=PC1, y=PC2, color=Phenotype), alpha=0.75)+
  scale_color_manual(values=c("red","blue"))+
  theme_bw()+
  theme(panel.grid=element_blank())

#poptable<-poppr(gen)
#save(poptable, file="outputs/500/poptable.rda")
load("outputs/500/poptable.rda")
poptable

```


## Separating Phenotypes Based on Genetic Data - DAPC
A principal component analysis shows the axes of maximum variation in the data (Note: While writing this I realized I did not examine PC's 3 or 4 as these were not calculated by the function I used. I will examine those in the future). We have here two groups, and, we can maximize separation of the two groups using a discriminant analysis of principal components (DAPC). As we have only two groups and no strong signal for more than that, a single discriminant function is shown.

```{r, warning=FALSE, message=FALSE, echo=FALSE}

dapc<-dapc(gen, n.pca=175, n.da=1)

#Red is FWR, Blue is MIG.    
scatter(dapc, col=c("red","blue"))
```


This approach separates our groups (FWR=red, MIG=blue) with individual positions shown in the carpet plot on the x-axis, density on the y-axis. The slightly bimodal density plots can be examined by looking at the posterior probability of assignment into the FWR and MIG groups. I have excluded all fish with a posterior probability > 0.95, resulting in 14 individuals. Notably, there are individuals in each group that have a 0 posterior probability of assignment to their prior phenotypic class, these are:
2013-SKT-6421    				
2013-SKT-6261   

```{r, warning=FALSE, message=FALSE, echo=FALSE}

post<-as_tibble(dapc$posterior)
post$Phenotype<-dapc$grp

ggplot((post %>% filter(Phenotype == "FWR" & FWR < 0.95)), aes(x=FWR))+
  geom_histogram(color="darkred", fill="red")+
  ggtitle("Posterior Probability of FWR Assigned to FWR")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))+
  theme(panel.grid = element_blank())

ggplot((post %>% filter(Phenotype == "MIG" & MIG < 0.95)), aes(x=MIG))+
  geom_histogram(color="darkblue", fill="blue")+
  ggtitle("Posterior Probability of MIG Assigned to MIG")+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5))+
  theme(panel.grid = element_blank())

#samples<-read_tsv("outputs/100/meta2.rda")
#samples %>% filter(`Sequence File Name` %in% c("Ht19-31_2012_G04","Ht20-30_2012_F04")) %>% select(`Sample_ID`)

```
What I find particularly interesting about this is that there are a lot of values around 0.50 and 0.75. These values are suggestive of hybrids.    

With DAPC, I can examine which variants contribute the most to the separation of phenotypes with a loading plot, and then pull out some allele frequencies.     


```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=5}
contrib <- loadingplot(dapc$var.contr, axis=1,
thres=.0017, lab.jitter=0)

tab(genind2genpop(genind[loc=c("contig_103_481573")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_571477")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_58_2863421")]),freq=TRUE)
tab(genind2genpop(genind[loc=c("contig_105394_1469")]),freq=TRUE)

#temp<-dapc$var.contr %>% as_tibble()
#temp$variant<-row.names(dapc$var.contr)
#c58<-temp %>% mutate(Chrom=gsub("_\\d+\\.\\d","",temp$variant)) %>% mutate(SNP=gsub("\\.\\d","",temp$variant)) %>% select(Chrom, SNP, LD1) %>% mutate(Loading=round(LD1, 8)) %>% select(-LD1) %>% unique() %>% filter(Chrom=="contig_58") 
#ggplot(c58)+geom_col(aes(x=SNP, y=Loading))
```

## Genome Wide Association Study
I conducted two different genome wide association studies on the data (ANGSD and PLINK). These methods are largely based around using a chi-squared test, and, often a p-value of 5 x 10-8 is considered meaningful to reduce false positives. Neither test did find significant results in the conventional sense based on a very small p-value. A lack of significant results based on a 5 x 10-8 cutoff has many precedents in the literature with several strategies on how to address it. Here are the top results from the ANGSD association test, distribution on different contigs, and a Manhattan style plot of a couple contigs.   
To be explicit, our p - values convert like so    
 -log10(0.01) = 2   
 -log10(0.001) = 3    
 -log10(0.0001) = 4   

```{r, warning=FALSE, message=FALSE, echo=FALSE}
data<-read_tsv(file="outputs/200/216.lrt0.gz")

#Converting to log10p
data <- data %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) 

# To be explicit, our p - values convert like so
# -log10(0.01) = 2
# -log10(0.001) = 3
# -log10(0.0001) = 4


#Check for most significant values
data %>% filter(log10p > 4) %>% arrange(-log10p)

data %>% filter(log10p > 3) %>% group_by(Chromosome) %>% summarize(Count=n()) %>% arrange(-Count)

#read_tsv(file="outputs/300/202.lrt0.gz") %>% arrange(-LRT) %>% mutate(log10p=-log10(dchisq(LRT, df = 1))) %>% filter(Chromosome==58)
ggplot(data%>%filter(Chromosome %in% c(58,114487,62,103) & log10p>=0.5))+
  geom_point(aes(x=Position, y=log10p, color=log10p), alpha=0.95, cex=1.5)+
  geom_hline(yintercept =4, size=0.5, alpha=0.5)+
  facet_wrap(~ Chromosome, ncol=4, scales="free_x")+
  scale_color_gradient(low="gray", high="skyblue")+
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1))+
  theme(panel.grid = element_blank())
```

Considering the top 1% of results shared between separate tests has been reported in the literature as a method of determining which SNPs may be associated with the trait of interest. This approach highlights 43 SNPs for me, mostly found on the assembly contigs 114514 and 58. This includes top SNPs shown by DAPC to be able to separate the two phenotypes genetically.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(ggrepel)
plink<-read_table(file = "outputs/300/plink-gwas.assoc.adjusted")

#Getting the top 1%
top1<-plink %>% top_frac(-.01, GC)

data<-read_tsv(file="outputs/200/216.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>% mutate(SNP = paste0("contig_",Chromosome,"_",Position))
top2<-data %>% top_frac(.01, log10p)

comb<-inner_join(top1, top2)

text<-comb%>%filter(Chromosome==58)%>% mutate(Text=gsub("contig_","",SNP))
total<-inner_join(plink,data)
ggplot(total, aes(x=log10p, y=-log10(GC)))+geom_point(color="darkgrey",alpha=0.5)+
  geom_point(data=comb, aes(x=log10p, y=-log10(GC)), color="red", alpha=0.75, cex=1)+
  geom_point(data=text, aes(x=log10p, y=-log10(GC)), color="black", alpha=0.75, cex=1.5)+
  geom_text_repel(data=text, aes(x=log10p, y=-log10(GC), label=Text ) )+
  ylab("-log10p PLINK")+
  xlab("-log10p ANGSD")+
  theme_bw()+
  theme(panel.grid = element_blank())+
  ggtitle("")

#Where are most of these found?
comb %>% group_by(Chromosome) %>% summarize(Count=n()) %>% arrange(-Count)
```

# What to do next?
Examining numerous prinicpal components (PC1-PC6) while keeping in mind date, sampling locations and other variables. Another important aspect to consider is the distribution of linkage blocks, especially on those contigs with stronger associations with phenotype. I have also begun a random forest analysis to create a predictive method to assign fish to phenotype based on their genetic background.       


